<!DOCTYPE html><html lang="zh-CN"><head><title>By Date | MountAye·Blog</title><meta property="og:title" content="By Date | MountAye·Blog"/><meta name="twitter:title" content="By Date | MountAye·Blog"/><meta name="description" content="Intellectual thinking is a lifestyle."/><meta property="og:description" content="Intellectual thinking is a lifestyle."/><meta name="twitter:description" content="Intellectual thinking is a lifestyle."/><meta property="og:image" content="https://blog.mountaye.com/assets/img/before_h2.png"/><meta name="twitter:image" content="https://blog.mountaye.com/assets/img/before_h2.png"/><meta charSet="utf-8"/><meta http-equiv="X-UA-Compatible" content="IE=edge"/><meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no"/><meta name="generator" content="Next.js v14.2.6"/><link rel="canonical" href="https://blog.mountaye.com/en/history"/><link rel="icon" href="/favicon.ico" type="image/x-icon"/><meta property="og:locale" content="en"/><meta property="og:url" content="https://blog.mountaye.com/en/history"/><meta name="twitter:card" content="summary_large_image"/><meta name="next-head-count" content="17"/><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.8/dist/katex.min.css"/><link rel="preload" href="/_next/static/css/70a087ca0a2aa809.css" as="style"/><link rel="stylesheet" href="/_next/static/css/70a087ca0a2aa809.css" data-n-g=""/><link rel="preload" href="/_next/static/css/e94e0c2dba26aa75.css" as="style"/><link rel="stylesheet" href="/_next/static/css/e94e0c2dba26aa75.css" data-n-p=""/><noscript data-n-css=""></noscript><script defer="" nomodule="" src="/_next/static/chunks/polyfills-42372ed130431b0a.js"></script><script src="/_next/static/chunks/webpack-2df7a8d27de1794c.js" defer=""></script><script src="/_next/static/chunks/framework-64ad27b21261a9ce.js" defer=""></script><script src="/_next/static/chunks/main-e618b1edbb350739.js" defer=""></script><script src="/_next/static/chunks/pages/_app-82a2279cdebeb61d.js" defer=""></script><script src="/_next/static/chunks/24-03c6412f1cc555e5.js" defer=""></script><script src="/_next/static/chunks/923-376360315f0822f9.js" defer=""></script><script src="/_next/static/chunks/pages/en/history-a1d62021b92c4663.js" defer=""></script><script src="/_next/static/buildID/_buildManifest.js" defer=""></script><script src="/_next/static/buildID/_ssgManifest.js" defer=""></script></head><body><div id="__next"><div id="all" class="w-full h-max"><nav id="top" class="w-full h-11 sm:h-12 lg:h-14 fixed top-0 flex flex-row justify-between items-center z-50 bg-white dark:bg-black shadow-lg dark:shadow-gray-700 transition-transform"><a class="whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:text-slate-400 text-primary underline-offset-4 hover:underline h-10 py-2 w-max px-2 flex flex-row justify-end items-center grow-0 shrink-0" href="/en"><span class="relative flex shrink-0 overflow-hidden rounded-full h-9 w-9 sm:h-10 sm:w-10 md:h-12 md:w-12"><span class="flex h-full w-full items-center justify-center rounded-full bg-muted">Aye</span></span><div id="site title" class="pl-2 hidden sm:block"><span class="text-2xl font-serif dark:font-sans">MountAye</span><span class="ml-[0.5em] text-base font-sans hidden lg:inline">Blog</span></div></a><div class=" grow-0 shrink flex w-[25%]"><input type="text" class="flex h-10 w-full rounded-md border border-input px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 rounded-r-none bg-inherit dark:border-gray-700" id="search-input" placeholder="Search"/><button class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:text-slate-400 border border-input hover:bg-accent hover:text-accent-foreground h-10 w-10 rounded-none border-x-0 bg-inherit dark:border-gray-700" type="button" id="radix-:R19km:" aria-haspopup="menu" aria-expanded="false" data-state="closed"><span class="relative flex shrink-0 overflow-hidden h-5 w-5 rounded-none"><span class="flex h-full w-full items-center justify-center rounded-full bg-muted">Aye</span></span></button><button class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:text-slate-400 border border-input hover:bg-accent hover:text-accent-foreground h-10 w-10 rounded-l-none bg-inherit dark:border-gray-700"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-search h-4 w-4"><circle cx="11" cy="11" r="8"></circle><path d="m21 21-4.3-4.3"></path></svg></button></div><div class="flex flex-row"><button class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:text-slate-400 hover:bg-accent hover:text-accent-foreground h-10 py-2 md:hidden px-1 md:px-2 lg:px-3" type="button" id="radix-:R2tkm:" aria-haspopup="menu" aria-expanded="false" data-state="closed"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-layers"><path d="m12.83 2.18a2 2 0 0 0-1.66 0L2.6 6.08a1 1 0 0 0 0 1.83l8.58 3.91a2 2 0 0 0 1.66 0l8.58-3.9a1 1 0 0 0 0-1.83Z"></path><path d="m22 17.65-9.17 4.16a2 2 0 0 1-1.66 0L2 17.65"></path><path d="m22 12.65-9.17 4.16a2 2 0 0 1-1.66 0L2 12.65"></path></svg></button><div class="hidden md:block "><a class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:text-slate-400 hover:bg-accent hover:text-accent-foreground h-10 py-2 px-1 md:px-2 lg:px-3" href="/en/history"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-history"><path d="M3 12a9 9 0 1 0 9-9 9.75 9.75 0 0 0-6.74 2.74L3 8"></path><path d="M3 3v5h5"></path><path d="M12 7v5l4 2"></path></svg></a><a class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:text-slate-400 hover:bg-accent hover:text-accent-foreground h-10 py-2 px-1 md:px-2 lg:px-3" href="/en/topics"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-tags"><path d="m15 5 6.3 6.3a2.4 2.4 0 0 1 0 3.4L17 19"></path><path d="M9.586 5.586A2 2 0 0 0 8.172 5H3a1 1 0 0 0-1 1v5.172a2 2 0 0 0 .586 1.414L8.29 18.29a2.426 2.426 0 0 0 3.42 0l3.58-3.58a2.426 2.426 0 0 0 0-3.42z"></path><circle cx="6.5" cy="9.5" r=".5" fill="currentColor"></circle></svg></a><a class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:text-slate-400 hover:bg-accent hover:text-accent-foreground h-10 py-2 px-1 md:px-2 lg:px-3" href="/en/discuss"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-messages-square"><path d="M14 9a2 2 0 0 1-2 2H6l-4 4V4c0-1.1.9-2 2-2h8a2 2 0 0 1 2 2z"></path><path d="M18 9h2a2 2 0 0 1 2 2v11l-4-4h-6a2 2 0 0 1-2-2v-1"></path></svg></a></div><button class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:text-slate-400 hover:bg-accent hover:text-accent-foreground h-10 py-2 px-1 md:px-2 lg:px-3" type="button" id="radix-:R1dkm:" aria-haspopup="menu" aria-expanded="false" data-state="closed"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-languages w-5 h-5"><path d="m5 8 6 6"></path><path d="m4 14 6-6 2-3"></path><path d="M2 5h12"></path><path d="M7 2h1"></path><path d="m22 22-5-10-5 10"></path><path d="M14 18h6"></path></svg><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-chevron-down w-4 h-4 hidden md:block"><path d="m6 9 6 6 6-6"></path></svg></button><div class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:text-slate-400 hover:bg-accent hover:text-accent-foreground h-10 py-2 px-1 md:px-2 lg:px-3 flex items-center space-x-1"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-sun"><circle cx="12" cy="12" r="4"></circle><path d="M12 2v2"></path><path d="M12 20v2"></path><path d="m4.93 4.93 1.41 1.41"></path><path d="m17.66 17.66 1.41 1.41"></path><path d="M2 12h2"></path><path d="M20 12h2"></path><path d="m6.34 17.66-1.41 1.41"></path><path d="m19.07 4.93-1.41 1.41"></path></svg><button type="button" role="switch" aria-checked="false" data-state="unchecked" value="off" class="peer inline-flex h-6 w-11 shrink-0 cursor-pointer items-center rounded-full border-2 border-transparent transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 focus-visible:ring-offset-background disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=unchecked]:bg-input" id="night-mode"><span data-state="unchecked" class="pointer-events-none block h-5 w-5 rounded-full bg-background shadow-lg ring-0 transition-transform data-[state=checked]:translate-x-5 data-[state=unchecked]:translate-x-0 dark:bg-slate-500"></span></button><input type="checkbox" aria-hidden="true" style="transform:translateX(-100%);position:absolute;pointer-events:none;opacity:0;margin:0" tabindex="-1" value="off"/><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-moon text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 hidden md:block" for="night-mode"><path d="M12 3a6 6 0 0 0 9 9 9 9 0 1 1-9-9Z"></path></svg></div></div></nav><aside id="side" class="
                        hidden md:flex flex-col 
                        w-36 h-screen
                        fixed left-0 bottom-0 
                        z-40 bg-white dark:bg-black shadow-2xl dark:shadow-gray-700         
             "><div class="h-16 grow-0"></div><div class="grow py-20 px-4 text-base font-serif dark:font-sans text-wrap overflow-x-auto utils_right2left__q5Pl7 utils_verticalLines__s6Q8n"><p id="quote-line">智力活动是一种生活态度</p><p>——<span id="quote-author">阿掖山，一个博客</span></p></div><footer class="py-1 grow-0 shrink-0 flex flex-col items-center"><div class="text-sm"><p>MountAye<!-- --> © <!-- -->2024<br/>Rights Reserved</p></div><div class="flex flex-row"><a class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:text-slate-400 hover:bg-accent hover:text-accent-foreground h-10 py-2 px-1" href="/en/feed.xml"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-rss stroke-[#3c70c6]"><path d="M4 11a9 9 0 0 1 9 9"></path><path d="M4 4a16 16 0 0 1 16 16"></path><circle cx="5" cy="19" r="1"></circle></svg></a><a class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:text-slate-400 hover:bg-accent hover:text-accent-foreground h-10 py-2 px-1" href="https://www.github.com/MountAye"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-github stroke-[#3c70c6]"><path d="M15 22v-4a4.8 4.8 0 0 0-1-3.5c3 0 6-2 6-5.5.08-1.25-.27-2.48-1-3.5.28-1.15.28-2.35 0-3.5 0 0-1 0-3 1.5-2.64-.5-5.36-.5-8 0C6 2 5 2 5 2c-.3 1.15-.3 2.35 0 3.5A5.403 5.403 0 0 0 4 9c0 3.5 3 5.5 6 5.5-.39.49-.68 1.05-.85 1.65-.17.6-.22 1.23-.15 1.85v4"></path><path d="M9 18c-4.51 2-5-2-7-2"></path></svg></a><a class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:text-slate-400 hover:bg-accent hover:text-accent-foreground h-10 py-2 px-1" href="https://www.twitter.com/MountAye"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-twitter stroke-[#3c70c6]"><path d="M22 4s-.7 2.1-2 3.4c1.6 10-9.4 17.3-18 11.6 2.2.1 4.4-.6 6-2C3 15.5.5 9.6 3 5c2.2 2.6 5.6 4.1 9 4-.9-4.2 4-6.6 7-3.8 1.1 0 3-1.2 3-1.2z"></path></svg></a><a class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:text-slate-400 hover:bg-accent hover:text-accent-foreground h-10 py-2 px-1" href="https://github.com/MountAye/comments/discussions/categories/blog"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-mail stroke-[#3c70c6]"><rect width="20" height="16" x="2" y="4" rx="2"></rect><path d="m22 7-8.97 5.7a1.94 1.94 0 0 1-2.06 0L2 7"></path></svg></a></div></footer></aside><div id="main" class="
                    block h-full w-screen pt-16
                    sm:w-[640px] sm:ml-auto sm:mr-0 
                    md:w-[calc(100vw-144px)]
         "><div class="rounded-lg bg-card text-card-foreground pt-8 shadow-lg border-0 dark:shadow-gray-700"><div class="p-6 pt-0"><div class="relative w-full overflow-auto"><table class="w-full caption-bottom text-sm text-nowrap overflow-hidden"><thead class="[&amp;_tr]:border-0 border-0 shadow-md dark:shadow-gray-700"><tr class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"><th class="h-12 px-4 text-left align-middle text-muted-foreground [&amp;:has([role=checkbox])]:pr-0 font-bold">Date</th><th class="h-12 px-4 text-left align-middle text-muted-foreground [&amp;:has([role=checkbox])]:pr-0 font-bold">Title</th><th class="h-12 px-4 text-left align-middle text-muted-foreground [&amp;:has([role=checkbox])]:pr-0 font-bold">Tags</th><th class="h-12 px-4 text-left align-middle text-muted-foreground [&amp;:has([role=checkbox])]:pr-0 font-bold">Excerpt</th></tr></thead><tbody class="[&amp;_tr:last-child]:border-0"><tr class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted dark:border-gray-700"><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">2024-05-14</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0"><a class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:text-slate-400 underline-offset-4 hover:underline h-10 px-4 py-2 pl-0 text-[#3c70c6]" href="/en/articles/information-entropy-kl-divergence-cross-entropy-mutual-information"><b>.tex | Compare 2 Probability Distribution or 2 Messages</b></a></td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 font-mono"><a style="color:#329894;font-weight:bold;margin-right:0.5em" href="/en/topics#tex">.<!-- -->tex</a><a style="color:#3c70c6;font-weight:bold;margin-right:0.5em" href="/en/topics#phy">.<!-- -->phy</a><a style="color:#de2214;font-weight:bold;margin-right:0.5em" href="/en/topics#m">.<!-- -->m</a></td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">Translated by ChatGPT.</td></tr><tr class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted dark:border-gray-700"><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">2023-10-16</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0"><a class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:text-slate-400 underline-offset-4 hover:underline h-10 px-4 py-2 pl-0 text-[#3c70c6]" href="/en/articles/python-test-notes"><b>.py | Python Testing Notes</b></a></td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 font-mono"><a style="color:#ffbf00;font-weight:bold;margin-right:0.5em" href="/en/topics#md">.<!-- -->md</a><a style="color:#4584b6;font-weight:bold;margin-right:0.5em" href="/en/topics#py">.<!-- -->py</a></td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">Translated by ChatGPT.</td></tr><tr class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted dark:border-gray-700"><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">2023-10-13</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0"><a class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:text-slate-400 underline-offset-4 hover:underline h-10 px-4 py-2 pl-0 text-[#3c70c6]" href="/en/articles/python-packaging"><b>.py | Making Your Code Usable by Others</b></a></td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 font-mono"><a style="color:#ffbf00;font-weight:bold;margin-right:0.5em" href="/en/topics#md">.<!-- -->md</a><a style="color:#4584b6;font-weight:bold;margin-right:0.5em" href="/en/topics#py">.<!-- -->py</a></td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">Translated by ChatGPT.</td></tr><tr class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted dark:border-gray-700"><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">2023-05-19</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0"><a class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:text-slate-400 underline-offset-4 hover:underline h-10 px-4 py-2 pl-0 text-[#3c70c6]" href="/en/articles/OPT-workshop-notes"><b>.doc | OPT Workshop Notes</b></a></td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 font-mono"><a style="color:#285191;font-weight:bold;margin-right:0.5em" href="/en/topics#doc">.<!-- -->doc</a></td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">OPT is the abbreviation of Optional Practice Training.</td></tr><tr class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted dark:border-gray-700"><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">2023-04-05</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0"><a class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:text-slate-400 underline-offset-4 hover:underline h-10 px-4 py-2 pl-0 text-[#3c70c6]" href="/en/articles/job-search-efforts-during-march-meeting"><b>.doc | Job Search Efforts During March Meeting</b></a></td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 font-mono"><a style="color:#285191;font-weight:bold;margin-right:0.5em" href="/en/topics#doc">.<!-- -->doc</a></td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">[Translated by ChatGPT] 2 interviews and 1 focus session during March Meeting 2023.</td></tr><tr class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted dark:border-gray-700"><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">2023-03-12</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0"><a class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:text-slate-400 underline-offset-4 hover:underline h-10 px-4 py-2 pl-0 text-[#3c70c6]" href="/en/articles/notes-during-APS-march-meeting-2022-Chicago"><b>.tex | Notes during APS March Meeting 2022 in Chicago</b></a></td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 font-mono"><a style="color:#329894;font-weight:bold;margin-right:0.5em" href="/en/topics#tex">.<!-- -->tex</a><a style="color:#3c70c6;font-weight:bold;margin-right:0.5em" href="/en/topics#phy">.<!-- -->phy</a></td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">Notes during the March Meeting in 2022 Chicago</td></tr><tr class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted dark:border-gray-700"><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">2023-03-12</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0"><a class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:text-slate-400 underline-offset-4 hover:underline h-10 px-4 py-2 pl-0 text-[#3c70c6]" href="/en/articles/notes-during-APS-march-meeting-2023-Las-Vegas"><b>.tex | Notes during APS March Meeting 2023 in Las Vegas</b></a></td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 font-mono"><a style="color:#329894;font-weight:bold;margin-right:0.5em" href="/en/topics#tex">.<!-- -->tex</a><a style="color:#3c70c6;font-weight:bold;margin-right:0.5em" href="/en/topics#phy">.<!-- -->phy</a></td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">Notes during APS March Meeting 2023 in Las Vegas</td></tr><tr class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted dark:border-gray-700"><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">2022-07-08</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0"><a class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:text-slate-400 underline-offset-4 hover:underline h-10 px-4 py-2 pl-0 text-[#3c70c6]" href="/en/articles/python-subprocess-run-bash"><b>.py | Executing Bash Commands with python.subprocess</b></a></td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 font-mono"><a style="color:#ffbf00;font-weight:bold;margin-right:0.5em" href="/en/topics#md">.<!-- -->md</a><a style="color:#4584b6;font-weight:bold;margin-right:0.5em" href="/en/topics#py">.<!-- -->py</a></td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">Translated by ChatGPT.</td></tr><tr class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted dark:border-gray-700"><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">2021-12-11</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0"><a class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:text-slate-400 underline-offset-4 hover:underline h-10 px-4 py-2 pl-0 text-[#3c70c6]" href="/en/articles/python-import-script-module-package"><b>.py | Import Existing Code</b></a></td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 font-mono"><a style="color:#ffbf00;font-weight:bold;margin-right:0.5em" href="/en/topics#md">.<!-- -->md</a><a style="color:#4584b6;font-weight:bold;margin-right:0.5em" href="/en/topics#py">.<!-- -->py</a></td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">Translated by ChatGPT.</td></tr><tr class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted dark:border-gray-700"><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">2021-10-13</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0"><a class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:text-slate-400 underline-offset-4 hover:underline h-10 px-4 py-2 pl-0 text-[#3c70c6]" href="/en/articles/install-pytorch-cpu-on-fedora"><b>.py | Installing the CPU Version of PyTorch on Fedora</b></a></td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 font-mono"><a style="color:#ffbf00;font-weight:bold;margin-right:0.5em" href="/en/topics#md">.<!-- -->md</a><a style="color:#4584b6;font-weight:bold;margin-right:0.5em" href="/en/topics#py">.<!-- -->py</a><a style="color:#708090;font-weight:bold;margin-right:0.5em" href="/en/topics#ai">.<!-- -->ai</a></td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">Translated by ChatGPT.</td></tr><tr class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted dark:border-gray-700"><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">2021-06-29</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0"><a class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:text-slate-400 underline-offset-4 hover:underline h-10 px-4 py-2 pl-0 text-[#3c70c6]" href="/en/articles/python-interpreter-editor-virtualenv"><b>.py | Python&#x27;s Interpreter, Editor, and Virtual Environment</b></a></td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 font-mono"><a style="color:#ffbf00;font-weight:bold;margin-right:0.5em" href="/en/topics#md">.<!-- -->md</a><a style="color:#4584b6;font-weight:bold;margin-right:0.5em" href="/en/topics#py">.<!-- -->py</a></td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">Really, what are Python&#x27;s interpreter, editor, and virtual environment</td></tr><tr class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted dark:border-gray-700"><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">2021-05-15</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0"><a class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:text-slate-400 underline-offset-4 hover:underline h-10 px-4 py-2 pl-0 text-[#3c70c6]" href="/en/articles/mpipks-non-equilibrium-physics-note-10-dimension-reduction"><b>.mpipks-note | 10. Dimension Reduction</b></a></td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 font-mono"><a style="color:#000000;font-weight:bold;margin-right:0.5em" href="/en/topics#mpipks-note">.<!-- -->mpipks-note</a></td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">As the title suggests</td></tr><tr class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted dark:border-gray-700"><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">2021-03-07</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0"><a class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:text-slate-400 underline-offset-4 hover:underline h-10 px-4 py-2 pl-0 text-[#3c70c6]" href="/en/articles/python-installation-and-configuration"><b>.py | Setup a Python Dev Environment</b></a></td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 font-mono"><a style="color:#ffbf00;font-weight:bold;margin-right:0.5em" href="/en/topics#md">.<!-- -->md</a></td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">Setup a Python dev environment on a Windows machine, using barebone Python, virtualenvwrapper, vscode, and without Conda.</td></tr></tbody></table></div></div></div></div></div></div><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{"articles":[{"slug":"information-entropy-kl-divergence-cross-entropy-mutual-information","filename":"2024-05-14-information-entropy-kl-divergence-cross-entropy-mutual-information.md","date":"2024-05-14","title":".tex | Compare 2 Probability Distribution or 2 Messages","layout":"post","keywords":["tex","phy","m"],"excerpt":"Translated by ChatGPT.","content":"\n\u003e After feeling self-satisfied for a while, I realized that this article is basically a translation of the English Wikipedia entry Quantities of Information. However, the corresponding Chinese Wikipedia entry doesn’t cover as much content as the English version, so it’s not entirely a waste of effort.\n\n## Information and Probability\n\nInformation is expressed through a proposition, which can represent multiple propositions through logical operations. This proposition answers a specific question. Before receiving definite information, many other propositions could potentially be answers to this question, making probability theory applicable. Together, all possible propositions form a random variable space.\n\nFor example, in a multiple-choice question with four options, A, B, C, and D:\n- If it’s a single-choice question, the random variable space for the answer is $${A,B,C,D}$$.\n- If it’s multiple-choice, it’s $${A,B,C,D,AB,AC,AD,BC,BD,CD,ABC,ABD,ACD,BCD,ABCD}$$, and so on.\n\n## Quantifying Information in a Probability Distribution\n\n### Self-Information\n\nSelf-information is a property of a random event or a specific possible value of a random variable. It is expressed as:\n\n$$ I(m) = -\\log_n\\left(p(M=m)\\right) $$\n\nThis is a dimensionless quantity, but the logarithm base can vary:\n\n- When $$n=2$$, the unit is called a bit or shannon. A shannon represents the **maximum information** a binary digit can convey.\n- When $$n=e$$, the unit is nat, with loge or $$ln=\\log_e$$ being the natural logarithm.    \n- When $$n=10$$, the unit is hartley.\n\nThe formula is significant in information theory as it represents the minimum number of single-choice questions with nnn options required to find the answer.\n\n### Entropy\n\nEntropy is the overall property of a random variable’s probability distribution. The formula is straightforward, as it’s the expected value of self-information:\n\n$$S(p(M))=\\mathbb{E}_p[-\\log_n p(M)]=-\\sum_{m\\in M}p(m)\\log_n p(m)$$\n\nFor example, binary entropy for a specific random event would be:\n\n$$ S_{binary} = -(1-p)\\log(1-p)-p\\log p $$\n\nEntropy in information theory signifies the minimum number of questions required to equate to the information contained in multiple propositions.\n\n## Comparing Information Content in Two Probability Distributions\n\n### Kullback–Leibler (K-L) Divergence\n\nAlso known as relative entropy or I-divergence, K-L divergence is defined for two probability distributions of the same random variable space:\n\n$$ D_{KL}(p(X)|q(X))=\\sum_{x\\in X}p(x)\\log\\frac{p(x)}{q(x)} $$\n\nThis measures the **extra** number of single-choice questions required when using the optimal questions for $$q(X)$$ instead of $$p(X)$$.\n\n### Cross Entropy\n\nCross entropy describes the total number of single-choice questions needed when using the questions optimal for $$q(X)$$ with respect to $$p(X)$$:\n\n$$ CE(p(X),q(X))=-\\sum_{x\\in X}p(x)\\log q(x) $$\n\nBinary cross entropy between ppp and qqq can be expressed as:\n\n$$ BCE(p,q)=-p\\log q-(1-p)\\log(1-q) $$\n\n\n### Mutual Information\n\nMutual information, defined for two different random variable spaces, XXX and YYY, is:\n\n$$ MI(X,Y)=\\sum_{x,y}p(x,y)\\log\\frac{p(x,y)}{p(x)p(y)} $$\n\nThis measures the difference between the joint distribution of $$X$$ and $$Y$$ and the assumption that $$X$$ and $$Y$$ are independent, with $$MI=0$$ if they are independent.\n\n## Sidebar Discussion\n\n### Loss Functions in PyTorch\n\nSome relevant loss functions in PyTorch are:\n\n- `torch.nn.KLDivLoss`\n- `torch.nn.CrossEntropyLoss`\n- `torch.nn.BCELoss`\n- `torch.nn.BCEWithLogitsLoss`\n\n### Boltzmann’s Epitaph\n\n$$ S=k\\log W $$\n\nHere, $$S$$ is the thermodynamic entropy, $$k$$ is Boltzmann’s constant $$k_B$$, and $$W$$ is the number of microstates. In statistical mechanics, when all states are equally probable, information entropy and thermodynamic entropy differ only by $$k_B$$.\n\n### Fortune Telling and Low Information Content\n\nIn 451 AD, Attila invaded Roman territory and battled a coalition led by Aetius, including Theodoric, King of the Visigoths. Before the battle, Attila consulted a fortune teller who predicted the death of a king and the collapse of a state. Attila assumed Theodoric would fall, but it turned out that while Theodoric did die, the coalition defeated the Huns, and Attila’s ambitions were thwarted.\n\nThis story illustrates how fortune tellers employ low-information predictions that lead people to infer higher-information outcomes, playing on the difference in entropy between statements.\n"},{"slug":"python-test-notes","filename":"2023-10-16-python-test-notes.md","date":"2023-10-16","title":".py | Python Testing Notes","layout":"post","keywords":["md","py"],"excerpt":"Translated by ChatGPT.","content":"\n## Some Terms\n\n### Debugging and Testing\n\nDebugging is generally done by the code's author and is used to check the program's execution process for mismatches between thought process and implementation. Debugging code is usually written alongside the main program and includes error handling and logging. Simple debugging can use `print` and `assert` statements, while more complex programs can use the `logging` module.\n\nTesting, on the other hand, is generally done by a third party. The test code is separated from the program code, and the person writing the tests doesn’t need to understand the exact workings of the program—only whether the program produces the declared output given certain inputs.\n\n### Unit Testing\n\nUnit testing is a process that verifies the correctness of a module, function, or class.\n\nThe testing method involves writing a set of test cases, where the tester provides various inputs to the module to check if its output is correct, and also verifies if the program correctly identifies and handles invalid inputs by raising exceptions.\n\nPassing all unit tests doesn’t guarantee that the entire program is error-free, but failing any unit test definitely indicates a problem with the program.\n\n### Documentation Testing\n\nDocumentation provides an overview of a code’s functionality, often with examples that include sample code and expected output. This process is similar to unit testing, except that it's purely descriptive.\n\nDocumentation testing is an automated process that finds example code within the documentation, runs it, and compares the output with the documented results.\n\n### Integration Testing\n\nIntegration testing simulates user behavior by testing the interactions between various modules, ensuring that the program can work properly in a production environment.\n\n## Testing in Python\n\n### File Structure\n\nIn the previous article, I mentioned the `src-layout` structure, which is also discussed in an [official blog post on setuptools](https://blog.ionelmc.ro/2014/05/25/python-packaging/#the-structure). One advantage of this file structure is that the test folder is typically at the same level as `src/` instead of the package itself. This way, you need to install the package in a virtual environment before running the tests, avoiding potential discrepancies between test code and user-downloaded content.\n\n```\n\u003cproject_name\u003e \n├── LICENSE \n├── pyproject.toml \n├── README.md \n├── src/ \n│ └── \u003cpackage_name\u003e/ \n│ ├── init.py \n│ └── example.py \n└── tests/\n```\n\n\n### Unit Testing with `unittest`\n\n`unittest` is Python’s built-in unit testing library.\n\nA basic test script looks like this:\n\n```python\nimport unittest\n\nclass TestName1(unittest.TestCase):\n\n    def test_sum(self):\n        self.assertEqual(sum([1, 2, 3]), 6, \"Should be 6\")\n\n    def test_sum_tuple(self):\n        self.assertEqual(sum((1, 2, 2)), 6, \"Should be 6\")\n\nclass TestName2(unittest.TestCase):\n\n    def test_sum(self):\n        self.assertEqual(sum([1, 2, 3]), 6, \"Should be 6\")\n\n    def test_sum_tuple(self):\n        self.assertEqual(sum((1, 2, 2)), 6, \"Should be 6\")\n\nif __name__ == '__main__':\n    unittest.main()\n```\n\n- The test cases are wrapped in a class that inherits from `unittest.TestCase`.\n- All test method names begin with \"test\". Limited properties are testable as they use built-in methods prefixed with `self.`. Some of these are in the table below, with `[assertTrue(x)](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertTrue)` and `assertRaises(Exception)` being especially versatile.\n- When run as `'__main__'`, it executes the built-in `unittest.main()` function.\n- Running the script requires specifying the script filename in Python. If you install the `nose2` library, you can run `python -m nose2`, which will automatically discover and execute all tests. (`unittest` also seems to have an automatic discovery feature.)\n\n### Documentation Testing\n\nhttps://www.liaoxuefeng.com/wiki/1016959663602400/1017605739507840\n\nPython's documentation testing uses the `doctest` library, as shown below:\n\n```python\nclass Dict(dict):\n    '''\n    Simple dict but also supports access as x.y style.\n\n    \u003e\u003e\u003e d1 = Dict()\n    \u003e\u003e\u003e d1['x'] = 100\n    \u003e\u003e\u003e d1.x\n    100\n    \u003e\u003e\u003e d1.y = 200\n    \u003e\u003e\u003e d1['y']\n    200\n    \u003e\u003e\u003e d2 = Dict(a=1, b=2, c='3')\n    \u003e\u003e\u003e d2.c\n    '3'\n    \u003e\u003e\u003e d2['empty']\n    Traceback (most recent call last):\n        ...\n    KeyError: 'empty'\n    \u003e\u003e\u003e d2.empty\n    Traceback (most recent call last):\n        ...\n    AttributeError: 'Dict' object has no attribute 'empty'\n    '''\n    def __init__(self, **kw):\n        super(Dict, self).__init__(**kw)\n\n    def __getattr__(self, key):\n        try:\n            return self[key]\n        except KeyError:\n            raise AttributeError(r\"'Dict' object has no attribute '%s'\" % key)\n\n    def __setattr__(self, key, value):\n        self[key] = value\n\nif __name__ == '__main__':\n    import doctest\n    doctest.testmod()\n```\n\nWith `docstring` indentation preserved, lines beginning with `\u003e\u003e\u003e` are treated as tests, with the following lines as expected outputs. For expected errors, you can use `...` to omit intermediate error details.\n\nIntegration Testing\n\nYou can also use unittest for integration testing.\n\nThe main difference from unit testing is that integration tests often involve setting up a test dataset. This requires overriding unittest.TestCase.setup().\n\n```python\nclass TestComplexData(unittest.TestCase):\n    def setUp(self):\n        # load test data\n        self.app = App(database='fixtures/test_complex.json')\n\n    def test_customer_count(self):\n        self.assertEqual(len(self.app.customers), 10000)\n\n    def test_existence_of_customer(self):\n        customer = self.app.get_customer(id=9999)\n        self.assertEqual(customer.name, u\"バナナ\")\n        self.assertEqual(customer.address, \"10 Red Road, Akihabara, Tokyo\")\n\nif __name__ == '__main__':\n    unittest.main()\n```\n"},{"slug":"python-packaging","filename":"2023-10-13-python-packaging.md","date":"2023-10-13","title":".py | Making Your Code Usable by Others","layout":"post","keywords":["md","py"],"excerpt":"Translated by ChatGPT.","content":"\n\u003e Here, \"others\" also includes the version of yourself six months later, who may no longer remember how this code was written.\n\nSome time ago, I wrote an article called [**Using Import to Reference Existing Code**](https://mountaye.github.io/blog/articles/python-import-script-module-package), which explained how to use code written by others. I briefly mentioned a module's `setup.py` file at the end but didn’t go further. Now, let’s continue.\n\n- [https://packaging.python.org/en/latest/tutorials/packaging-projects/](https://packaging.python.org/en/latest/tutorials/packaging-projects/)\n- [https://setuptools.pypa.io/en/latest/userguide/quickstart.html](https://setuptools.pypa.io/en/latest/userguide/quickstart.html)\n\n## Basic Process\n\n- Write your code and structure the project files in a specific way (see next section).\n- According to your project structure, fill in a `pyproject.toml`, `setup.cfg`, **or** `setup.py` **file**.\n- Install the `build` library, then run `python -m build` to create a `dist/` folder with your built files. Optionally, upload the project to PyPi or Conda.\n\n## Project File Structure\n\nCommon file structures include the src-layout and flat-layout. Some small projects consist of a single Python file.\n\n### src-layout\n\nIn the src-layout, the folder containing the package source code has another folder above it, typically named `src` (although it could be named differently). The `pyproject.toml` file is on the same level as the `src/` folder.\n\n```\n\u003cproject_name\u003e \n    ├── LICENSE \n    ├── pyproject.toml \n    ├── README.md \n    ├── src/ \n    │   └── \u003cpackage_name\u003e/\n    │   ├── init.py \n    │   └── example.py \n    └── tests/\n```\n\n### flat-layout\n\nIn a flat-layout, the folder containing the package source code is directly under the top level of the project, on the same level as `pyproject.toml`.\n\nThis structure is older and not generally recommended.\n\n```\n\u003cproject_name\u003e \n    ├── pyproject.toml # and/or setup.cfg/setup.py\n    ├── \u003cpackage_name\u003e \n    |   ├── init.py \n    |   └── ... (other Python files) \n    ├── test \n    |   └── ... (test files) \n    ├── README.md \n    └── LICENSE \n```\n\n\n### Single-File Project\n\nThis can be considered a special case of flat-layout.\n\n```\n\u003cproject_name\u003e \n    ├── pyproject.toml # and/or setup.cfg/setup.py \n    ├── \u003cmy_module\u003e.py \n    ├── README.rst or README.md \n    └── LICENSE\n```\n\n\n## Filling in `pyproject.toml`, `setup.cfg`, **or** `setup.py` **File**\n\nTo package your code into an installable package, at least one of these three files must appear in the project root.\n\nThe file must contain project information in the syntax specific to its extension, most of which is self-explanatory.\n\nKey parameters, such as `name`, `packages`, and `package_dir`, are based on the file structure. If the file structure fully follows the structure outlined in the previous section, then [automatic discovery](https://setuptools.pypa.io/en/latest/userguide/package_discovery.html#automatic-discovery) with `setuptools.find_packages()` should suffice.\n\n### `name`\n\nThis is a required parameter.\n\nNote: In the file structure from the previous section, there are two names: `\u003cproject_name\u003e` and `\u003cpackage_name\u003e`.\n\n- `\u003cproject_name\u003e` is the name of the development project as a whole. If you are using a version control system like git, this is the name of your repository.\n- `\u003cpackage_name\u003e` can—but doesn’t have to—be the name you import in other code. This import name is specified by the `name` parameter in `pyproject.toml` / `setup.cfg` **/** `setup.py`. It cannot contain hyphens and can only use underscores.\n\nIf your `name` parameter differs from `\u003cpackage_name\u003e`, you must also set the `package_dir` parameter.\n\nThere is also a third name: the one used for `pip install __` when uploaded to PyPI. This name can contain hyphens, like `scikit-image`.\n\n### `packages`\n\nThis parameter is a list, but it typically takes the result of `setuptools.find_packages()`.\n\nThe `find_packages()` function commonly accepts three optional parameters:\n\n- `where`: A path relative to `setup.py`\n- `include`: A list of glob patterns\n- `exclude`: A list of glob patterns\n\nLeaving all parameters blank uses automatic discovery.\n\n### `package_dir`\n\nThis parameter is a dictionary with two main uses:\n\n- In a standard src-layout, simply write `{\"\": \"src/\"}`, indicating that all code is within this folder.\n- If the Python module structure differs from the code file structure, use this dictionary to define the mapping between modules and folders.\n\nFile paths are relative to `setup.py`.\n\n### `py_modules`\n\nThis parameter is a list of file paths, mostly used for single-file structures.\n\n## Building and Uploading\n\nInstall the build tool: `python3 -m pip install --upgrade build`\n\nRun the build: `python3 -m build`\n\nThis will create a `dist/` folder containing the build files.\n\nTo allow others to install your program using `pip install`, you need to upload the build to PyPI. Instructions are [here](https://packaging.python.org/en/latest/tutorials/packaging-projects/#uploading-the-distribution-archives).\n\n## Installation\n\n### Static Installation\n\nPackages already uploaded to PyPI can be installed with `pip install \u003cpackage\u003e`. This is called static installation.\n\n### Dynamic Installation\n\nFor packages still under development, run `pip install -e .` in the directory containing `setup.py`. This is called dynamic installation because changes in the code are reflected in the referencing projects in real time.\n\n## Exercises\n\nHere is the file structure of a [data analysis project](https://gist.github.com/ericmjl/27e50331f24db3e8f957d1fe7bbbe510), slightly modified:\n\n```\n/path/to/project/directory/ \n    |-- notebooks/ \n        |-- 01-first-logical-notebook.ipynb \n        |-- 02-second-logical-notebook.ipynb \n        |-- prototype-notebook.ipynb \n        |-- archive/ \n            |-- no-longer-useful.ipynb \n    |-- src/ \n        |-- projectname/ \n            |-- init.py \n            |-- config.py \n            |-- data.py \n            |-- utils.py \n            |-- setup.py \n    |-- README.md \n    |-- data/ \n    |-- raw/ \n    |-- processed/ \n    |-- cleaned/ \n    |-- scripts/ \n        |-- script1.py \n        |-- script2.py \n        |-- archive/ \n            |-- no-longer-useful.py \n    |-- environment.yml\n```\n\nQuestions:\n\n1. Is this a flat-layout or src-layout?\n2. How should setup.py be written? What would be the output of `pwd` when executing a dynamic installation?\n"},{"slug":"OPT-workshop-notes","filename":"2023-05-19-OPT-workshop-notes.md","date":"2023-05-19","title":".doc | OPT Workshop Notes","layout":"post","keywords":["doc"],"excerpt":"OPT is the abbreviation of Optional Practice Training.","content":"\n## Who is eligible\n\n### Pre-completion\n\nBefore graduation, need to talk with advisor, not common\n\n### Post-completion\n\nafter graduation,\n\nhave entered the US, and have been in F1 status for 2 semesters\n\nor graduates except for dissertation \n\n- ABD: all but dissertation, finish before the end of EAD date\n- Otherwise need to go back home and complete the degree\n\n## OPT basics\n\nrelated to your field of study\n\nstill an F1 student, not a work visa.\n\n12 months OPT per level of education, not cumulative\n\neligible STEM degrees for extension of post-complete OPT, or pending H1B\n\nno job offer is required to apply. \n\nallows to change jobs and have multiple jobs\n\nemployer does not have to do anything, applicant is sponsoring themselves\n\ncan do paid/unpaid, this university cannot do unpaid work.\n\n## Only work after authorization\n\nnot eligible to work until EAD card\n\nno exception\n\nneed to create SEVP portal account after the OPT is approved and after the start date\n\nneed to report on change: US home address, employer and address, start and end date of employment\n\n## Do not accrue more than 90 days of unemployment\n\nwithin the authorization period on EAD card\n\n\u003e 20 hours to be considered employed\n\n## Keep document and F1 records updated\n\nrequest OPT-I20\n\n## Only enroll part-time in classes unrelated to the current or future field of study\n\nreach out to OISS advisor \n\n## Travel outside US\n\npossible with documents\n\n1. passport\n2. F1 visa\n3. I20 with signature\n4. EAD\n5. proof of employment (optional)\n\n**while OPT application pending is possible, but risky** because the EAD will be mailed to your US address\n\n## Verify admission status I94 record is correct\n\n[i94.cbp.dhs.gov](http://i94.cbp.dhs.gov) \n\n## summary\n\n![how to maintain the OPT status](/blog/photos/2023-05-19-main-status.png)\n\n## Who can apply for extension\n\n### 24-month STEM OPT extension\n\naply during the last 90 days end of OPT\n\nUSCIS must receive before the end\n\nmust have a job\n\nemployer must enroll in E-verify\n\nrequires a plan (Form I-983)\n\n### H1B petition pending or approved may requested a “cap gap”\n\ncap gap employer file H1B application on April 1\n\nif approved, H1B starts on October 1\n\nif OPT expires before Oct 1, “cap gap” extension is added in SEVIS until decision of H1B is decided\n\n## application process\n\n- workshop, get a packet\n- request OPT I20  form from OISS advisor\n- prepare and submit form I-765 OPT application to USCIS\n- **must be in US to apply for OPT**\n\n## application timeline\n\n**application estimated to take 4 weeks.**\n\n90 days before program end date → I-20 end date → 60 days after program end date\n\npick requested start date: 1 day after I 20 end date → end of 60 day grace period\n\n![time window of OPT application](/blog/photos/2023-05-19-time-window.png)\n\n## Packet\n\n### request OPT-I20\n\nrequested OPT start date form\n\nacademic advisor verification form `OR` ABD OPT form, one or the other\n\n### The packet is for getting OPT I-20 recommendation from OISS\n\n## OPT I-20\n\napplication to USCIS with the recommendation from OISS\n\n### evidence files to upload I-765  with OPT I-20\n\nimmigration docu: passport, I-94\n\nprevious OPT and CPT, can skip if you don’t have it\n\n2 by 2 inches passport style photo; taken within 30 days; Public library, Walgreens, and CVS offer the service\n\n### What if the EAD start date is after the end date of the grace period?\n\ndidn't quite hear the answer"},{"slug":"job-search-efforts-during-march-meeting","filename":"2023-04-05-job-search-efforts-during-march-meeting.md","date":"2023-04-05","title":".doc | Job Search Efforts During March Meeting","layout":"post","keywords":["doc"],"excerpt":"[Translated by ChatGPT] 2 interviews and 1 focus session during March Meeting 2023.","content":"\n### Before the Meeting\n\nThe March meeting will be held in Las Vegas this year.\n\nIt's about two weeks away from the conference, and I developed folliculitis at the base of my chin where I have a beard. It quickly worsened into herpes zoster, commonly known as \"internal heat\" in Chinese medicine. The second semester of the American academic year ends in May, and I haven't made much progress on my thesis. I can't find a job, and the application season for postdoctoral programs is in full swing with paper submissions. I watched helplessly as the application deadlines for two projects passed by. The postdoctoral project shared by a neighboring research group reposted their job advertisement online shortly after I submitted my application materials. I also applied to two research institutes in Germany, with one already sending a rejection letter and the other not responding at all. In addition to all this, I need to prepare for my presentation at the conference. It's no wonder I'm feeling stressed.\n\nOn Friday, we had a rehearsal for the group's presentations. I had initially planned to reuse the slides from last year's conference but realized that the main content from my paper was missing. It made me realize that I had indeed accomplished quite a bit over the past year. After receiving the rejection letter from the German research institute, I felt down for a day. However, the trend of using ChatGPT quickly gained popularity. So I compiled a list of professors I had heard of during my doctoral studies and fed their research group homepage descriptions, along with an abstract of my paper generated by AI, to ChatGPT, asking it to compose an email. The result was decent, and I received a reply inviting me to meet during the conference. The person who responded is a well-known researcher who studies animal behavior using a cross-scale approach and manages two research groups in Europe and East Asia. They made a big impression at last year's conference and later came to give a lecture at my university. We had lunch together. Shortly before the meeting, someone on the APS (American Physical Society) mailing list sent a mass email advertising a job opportunity. It turns out they are the current postdoc who will soon start a faculty position at a university in the western United States. Interestingly, they are currently working at the German research institute that rejected me. I can't help but feel a mix of envy and satisfaction knowing that they also fell for the ChatGPT-generated email trick.\n\n### Meeting\n\nI walked to the conference venue and checked in. I received a name tag, which was just a thick piece of paper hanging from a string, without even a plastic cover.\n\nAfter checking in, I surveyed the layout of the conference venue. The rooms related to biophysics were located on both sides of the first intersection in the main corridor. Each doorway had display boards indicating the themes of the speeches for each time slot. Unlike last year when I was the only one from our team attending, this year all three members of our team have presentations. In addition, Katie, a former master's student in our group, and her current classmates also have presentations. When I marked the schedule, I found that most of the interesting sessions conflicted with each other. Although I have social anxiety, I still feel that interpersonal relationships are important. Or perhaps it's because of my social anxiety that I feel interpersonal relationships are more important?\n\nBack at the hotel, I received a reply from my group email to confirm a time one or two weeks after the March meeting for a remote interview. The purpose is for me to introduce my current work and explain how I can contribute to his laboratory. It's a bit unexpected because his group email mentioned that he would attend the conference, and my email also stated that I was willing to meet during the conference. Does this reply imply a lack of interest? I feel a little disappointed.\n\n### Interview 1\n\nWhen I was wandering between different sub-venues and arrived late to one where the seats were already filled, I had to stand at the back. While I was standing and listening in one of the venues, a guy wearing a green jacket and carrying a sports backpack walked in. He listened for a while and then left. He seemed somewhat familiar, resembling the photos on ChatGPT's personal website.\n\nCurious, I discreetly left the venue and saw him talking to a friend in the corridor, with the name tag on the front of his chest turned over. I temporarily left and explored around, discovering that coffee was available in the main corridor. I lined up and got a cup, then came back to investigate and confirmed it was him.\n\nI greeted him, and it turned out that ChatGPT had been too verbose in his email and only mentioned that he was also attending the conference in March at the very end. I asked him when my presentation was scheduled and if he planned to attend, so we wouldn't need a Zoom meeting.\n\nThe title of my presentation during that time slot was quite peculiar, and it didn't seem to have much relevance to what I was working on. When I entered the waiting area, I noticed that apart from the two of us from our lab, all the other presentations in that session were focused on the same pattern and structure of a particular model organism. Our presentation came at the very end of the session, and I felt we performed well. When I started addressing the topic, I made a joke using the session's title, and we finished right on time. ChatGPT arrived at the venue about half an hour before my presentation, and after we finished, he left without coming up to talk to me.\n\nAfterward, back at the hotel, I received an email from him, inviting me to meet the next afternoon. I searched his schedule on the conference website and found that it was right after his presentation. Looking at the title of his talk, it seemed to be about his recent paper. I fed the article into an AI tool to generate an abstract but didn't get a chance to read it before my classmates pulled me away to play...\n\nHis presentation was scheduled for the next afternoon and fell in the middle of the overall time slot. I entered the venue at the beginning of the session, and he arrived a little later, sat for a while, and then left. The venue was twice as large as when I gave my presentation, but it wasn't completely filled. While he was speaking, I was reviewing the article summary generated by AI from yesterday, feeling that it contained more information than most of the presentations I had heard up until that point. His presentation skills were also far above the average level. His research targeted a biological system one level higher than my current research focus, but the approach and process seemed similar. His research was much more advanced, both in theory and experimentation. Although the theory seemed to be a fairly traditional model in the field, it was still much better than just presenting experimental data like we did.\n\nAfter the conference, we met in the corridor and found two seats. I showed him my slides, and he went through all the details, confirming that I would be working on gene editing in our lab's model organism. He also asked about the proportion of my work within the entire project.\n\nThen, he asked a question that felt as significant as being confessed to in a romantic relationship: \"What is the central question guiding your research career?\" Ah, my brain immediately crashed. Seeing me struggling, the guy said I could speak in Chinese, as he understood it but couldn't speak it. Finally, when I felt cornered, I told the truth - I just wanted to coast through academia, eat and wait for death. If\n\n ChatGPT has arrived, and we have super-intelligent AI, who needs physics anymore? Physics doesn't exist anymore. The guy chuckled at my response.\n\nHe then asked me to ask questions. I inquired about the duration of holidays, and he replied that it followed the school's policy. Those who have cultured cells should know that they might have to go to the school on weekends, so as long as the workload is sufficient, there are no attendance requirements on weekdays.\n\nOverall, the conversation went quite smoothly. In the end, he mentioned that he would send me an email and asked for two or three ways to contact me, as well as a recommendation letter.\n\n### Interview 2\n\nOriginally, I had a good impression of this esteemed person, but since it was the first day of the conference, I thought maybe everyone hadn't settled in yet. Therefore, I sent an email on Tuesday to confirm the meeting time, and as a result, we arranged to meet on Friday at noon, next to the DBIO service desk.\n\nI had planned to prepare in advance, as his research group had three or four presentations, but most of them clashed with my own presentation or the previous interview. In the end, I only attended one of them.\n\nFriday at noon turned out to be the last session of the entire conference. The DBIO table had already been removed, and everyone seemed eager to leave. After about ten minutes past 12 o'clock, I finally met the person, and five minutes later, the person I was talking to left. Everyone was busy packing up and closing the doors. We found an empty room with a table and chairs that had already been moved out. Inside the room, there was only one student lying on the floor sleeping. The two of us sat down facing each other on the floor.\n\nJust as I was about to take out my computer to go over my presentation, he waved his hand and asked me to briefly summarize my work orally. The subsequent series of questions and answers followed the same style, with the two of us sitting on the floor, exchanging rapid-fire questions and quick responses, resembling the agility of Zen. He started by asking why I chose to pursue my current work. I thought I gave a decent answer, but as I spoke, I realized that it was the answer from the previous interview. Why didn't I think of that earlier? Feeling a bit too pleased with myself, I forgot that this esteemed person was focused on studying behavior, not organizational structure. It could have been easily smoothed over, but I didn't handle it well. Then he asked about my favorite course during my undergraduate studies, a question he mentioned during our lunch break, so I pretended to be impressive. After that, he asked two significant questions: \"How did a stinky experimenter like you end up begging for food in our theoretical group?\" and \"How did a stinky structuralist like you end up begging for food in our behavioral research field?\" I didn't answer them well. Earlier, in my self-satisfaction, I mentioned having another job offer, which led to both sides being somewhat perfunctory during the latter part of the conversation.\n\nFinally, he briefed me that the research in East Asia has sufficient funding, although he spends a considerable amount of time there every year, it's not always possible to meet in person for guidance. Regarding Europe, if I decide to join, he would need to specifically apply for funding. The meaning behind this was quite clear, so we politely agreed to take some time to consider it and said we could stay in touch via email in the future.\n\n### Symposium\n\nThe symposium actually took place on the afternoon of the same day as my speech, right before the two interviews.\n\nThe so-called symposium was organized by several research institutions in Germany and aimed to promote information related to research conducted in Germany. It was quite far from the venues of the biophysics sessions. After my friend Katie finished her speech, I slipped out of the venue and rushed over. By the time I arrived, it had already begun. There were bagels, jam, and cheese on both sides of the entrance, and it was the first time in many days that I had lunch.\n\nThere were six or seven round tables in the venue, each with a small label indicating the name of a research institution. There were one or two staff members from each corresponding institute sitting at the tables. The host introduced each research institution in turn and gave the leaders of the institutions three to five minutes to introduce themselves. After that, it was free time. I noticed that the Max Planck Institute's table was near the bread, and since the other institutions didn't seem to focus much on basic scientific research, I found a chair and sat down. The head of the Max Planck Institute was an elder with white hair. His attire was more formal compared to the other participants, and he had a strong presence, resembling a wise figure.\n\nThe first question was about the difficulty of immigrating to Germany and whether Germany was somewhat exclusive. The elder replied that Germany is currently one of the most open countries in terms of immigration policies. Considering the phrase \"one of the,\" and also taking into account that foreign graduates in non-academic positions in the beacon country have to draw lots to obtain a work visa, excluding the \"borderline,\" this answer could not be considered wrong. Deutsche Welle had previously released a video with a similar statement, but that video had a follow-up phrase of \"aber...\" (but...).\n\nSomeone then asked about the projects and job structure at the Max Planck Institute. The answer was similar to what I had previously understood. Each research institution has the authority to recruit its own researchers. The duration of a postdoctoral position is generally two years, with extensions to three years or longer only when necessary. At the Max Planck Institute, except for the director, there are hardly any permanent positions for research positions. The institute serves as a stepping stone for researchers before they secure a permanent faculty position. However, whether it was a government regulation or something else, there was a rule stating that nine years after obtaining a doctoral degree, the employing institution must provide a permanent position. As a result, individuals are often dismissed before the end of those nine years. It wasn't asked whether this contradicted the previous answer.\n\nWhen asked what kind of research the Max Planck Institute favors, the elder explicitly stated that they prefer more challenging and ambitious topics rather than safe topics pursued solely to obtain a faculty position. He even specifically mentioned the aim to explore topics and scholars that have the potential to receive Nobel Prizes in the future. This naturally differs from the American-style notion of verbal freedom and the actual practice of focusing on safe bets. It's interesting how the understanding of pragmatism can differ so greatly between two cultures—one rooted in long-term plans and top-down designs, and the other more focused on short-term gains.\n\nSomeone also asked about the criteria for selecting applicants at the Max Planck Institute. The answer was that each research institution has its own criteria. However, he also mentioned his own criteria: academic performance is the main factor, and when conducting research, one should have a clear understanding of the underlying physics. So, how is academic performance judged? It mainly depends on attending prestigious schools, followed by academic rankings. If the school's name is less well-known, higher academic rankings are expected. The answer was sincere, and it didn't evoke any sense of objection towards such obvious discrimination based on one's undergraduate alma mater, nor did it provoke any intention to challenge it.\n\nThis response is clearly targeting the American model, which emphasizes research experience and places greater importance on the authorship of papers. The backbone of a research group consists of Ph.D. students who do not need to spend time attending classes. How much contribution can undergraduate students applying for a Ph.D. make to the entire project? Moreover, these contributions are largely made by sacrificing their efforts in coursework. The sentence highlighted above is a sarcastic comment on this matter. Furthermore, considering the notorious reputation of the academic environment in China, overseas research experience is typically required. Apart from special programs like the Everest Plan that provide scholarships, most students have to self-fund and contact foreign mentors. This not only turns academic selection into a disparity unrelated to academia, linked instead to students' economic conditions, but also fosters an intermediary industry that sells such information. What academic justice is there in this policy?\n\nHowever, let's consider how much academic performance can be reflected in a transcript. How many students openly ask their professors for grades when they intend to study abroad? Until a few years ago, there were even legitimate operations like \"passing and retaking\" courses at certain universities to alter transcripts. Not to mention the highly similar nature of examination questions across different subjects in China, which is detrimental to stability and unity. If attending prestigious schools is a requirement, then going abroad after high school graduation has far more advantages than going after undergraduate studies. It becomes a class-based filtering process, leading to the same destination through different paths.\n\nBoth of these scenarios involve individuals, each operating under a different moral framework, which is gradually deconstructed and distorted when translated into executable actions. Throughout this process, their own judgment is constantly outsourced. Outsourcing itself is not a problem, as division of labor is both a reflection of societal development and an important driving force. However, let's take an airplane as an example. Safely delivering passengers to their destinations requires coordination and trust between the crew and ground staff, but the safe arrival ultimately has objective standards. In the education and research system, the outcomes of its own cultivation are digested by itself. The selection process inherently involves the phenomenon of athletes acting as both players and referees. Some of the selected individuals may not be competent, while some of the non-selected ones may be capable. In that case, it would be straightforward: allocate all resources to the winners and cut off the paths for the losers...\n\nBoth models, in their mediocrity, peck at each other, and neither can replace the other. Ultimately, today's physics itself is lackluster. Leaving aside the magnificent breakthroughs during the birth of quantum mechanics and relativity, even the process of constructing the standard model was full of ad hoc solutions. Today, it is difficult to hope for its complete reversal. How much of the knowledge you solidly understand can be applied in your actual research? How many valuable works have you produced over these years of early involvement in research? Young people consider matters they believe are crucial to their future and destiny, but it is nothing more than a few bytes in their boss's email inbox. Whether they accept you or someone else, the difference is insignificant.\n\nIn short, there are no heroes in our time, which allows mediocre individuals to gain fame.\n\nIn the afternoon, a lady who had a conversation with \"群发哥\" (a nickname) the day before gave a speech shortly after him. Looking at her organization on the first page of the presentation slides, it seemed that she was a junior colleague in his doctoral laboratory and was about to join his current research institute. ~~In other words, she is the enemy who took away my offer~~. So, I posted a complaint in my friend circle. Her speech exceeded the allotted time and was interrupted by the host. I was about to\n\n add a comment saying \"Exceeded the time, so bad,\" but considering that I had been posting quite a lot on my friend circle in the past few days, I refrained.\n\nThe only person I admired during my four years in university, \"峰神\" (a nickname), replied below, saying he had also considered applying to the same research institute. I asked him where he is now, and a few minutes later, I saw his reply: \"Already started working.\" I quickly teased him, saying \"Rich brother, lend me 50 yuan.\" However, the words spoken at the meeting reverberated in my mind, sighing and turning into a deep breath. I silently put away my phone.\n"},{"slug":"notes-during-APS-march-meeting-2022-Chicago","filename":"2023-03-12-notes-during-APS-march-meeting-2022-Chicago.md","date":"2023-03-12","title":".tex | Notes during APS March Meeting 2022 in Chicago","layout":"post","keywords":["tex","phy"],"excerpt":"Notes during the March Meeting in 2022 Chicago","content":"\n## Physical of Artificial System\n\nsome physical learning that does not have to be in equilibrium\n\n## Cell fate\n\n### Xinhua Xing’s talk\n\nintro to differential geometry: Kamien, Rev Mod Phys 74:953 (2002)\n\nWeiKang Wang, in silico vector field analysis\n\nWaddington landscape 1957\n\nJoanathan Weissman Lab (WI, MIT) \n\n### Maria’s Talk\n\nBoston U, Pankaj Mahta Group.\n\nextension  of work from Lang et al PLoS comput bio (2014)\n\nfound an order parameter, better than UMAP \n\n### Shubham Tripathi’s talk\n\nmiimally frustrated topology\n\n### Glassy fluctuation ...\n\nproten and mRNA noise levels are Boltzmann distributed.\n\nspin glass: frustration\n\nfree energy \n\ngene expression noise can store information\n\n### Kraemer\n\ncell types in multicellular organisms\n\nmaster regulator as order parameter\n\n**Hopfield model**\n\nforward and backward conditional probabilities\n\n### Logan Carpenter\n\ntissue homeostasis: birth and death\n\ncellular Potts model\n\nthere is Hamiltonian!\n\nCompuCell3D software\n\ncell size distribution plot is very good\n\n### Lo’s talk\n\nUWashington\n\ncollision during cell cycle, replication and transcription\n\nF-mean\n\n### Lambert.io\n\nbacterial persistence\n\nCRISPER-dCas12a interference\n\nmassive parallel assay\n\n\n### Arkansas U talk\n\nReversibiliy and cell division dynamics of elongated Escherichia coli cells obtained at high pressure\n\na new cell division model\n\nGCDM\n\ncontinuous time Markov chain model?\n\n### Vivek\n\nmicrobial colonizing and phenotypes\n\nDanio rerio\n\nvirio Z20\n\n### Louis Cortes\n\nIPTG indeces lac operon\n\n$f(t)=1-e^{-\\frac{ln(2)}{t_{1/2}}(t-t_0)}$\n\n### Jintao Li\n\ncell competition\n\nhomeostasis pressure\n\n## Author \u0026 Reference Session\n\n### criteria\n\n- significantly advance physics\n- scientifically\n- letter\n    - advance in field\n    - open a new\n    - solve a problem\n    - be of great interest\n\n### for authors\n\n- identify audience\n- take-home message\n- adequate literature\n- authorship \u0026 acknowledgment\n- additional info\n    - suggested reviewer\n    - conflicts of interest\n    - related paper\n    - PRL/X: justification for suitability exclusively to\n\n## referee\n\n- novel, interesting, original\n- valid and reproducible\n- well presented\n- report\n    - summary\n    - technical details\n    - decision explained\n- report requirements\n    - substantiated\n    - a reasonable level of details\n    - no personal/self-serving remarks\n    - timeliness\n\n### joint submissions\n\n### updated PhySH\n\n## K42: Network Theory and Applications to Complex Systems\n\n### Complex Contagion: Unfolding and Control\n\nAdilson E. Motter, Northwest University\n\ncontagion\n\nwhack-a-mole effect\n\n### Physics of financial networks\n\nmono-layer networks: a network of ownership, a network of credit contracts between financial institutes\n\nmulti-layer  networks: \n\n## Non-equilibrium I\n\n### success landscape\n\n### speed fluctuation of bacterial replisomes\n\n### circadian clocks in natural environments\n\n$\\dot r = \\alpha r(1-\\frac{r^2}{R^2})$\n\n$\\dot \\theta = \\omega$\n\n“decoder”\n\nLeypunskiy 2017, Pittayakanchit 2018 \n\n## Physics of social interactions I\n\n### collective sensing in slime mold\n\nslime molds sense concentrations of fructose in environments\n\nhypotheses:\n\n- evenly distributed across the network. density be the only parameter\n- at the growing tips\n\n### [Competition for fluctuating resources reproduces statistics of species abundance over time across wide-ranging microbiotas](https://march.aps.org/sessions/M03/2)\n\nconsumer-resource model: species abundance is solely dependent on resource functuations\n\nTaylor’s law: distribution of abundance change is fit by a power law.\n\n“their model” coarse grained consumer-resource. can predict many kinds of , different resource competition regimes.\n\n### Dog Congnition\n\nAlexandra Horowitz. \n\nSocial play\n\n## Predicting Nonlinear and Complex Systems with Machine Learning I\n\n### Building Deep Learning Architectures for Physics, Chemistry, and Biology with Geometric Algebra\n\nfinished soon after I arrived\n\nvector institute\n\n### Quantum thermal machine with ML\n\nBraXXXX, VQE\n\nFinite time thermodynamics: compete between efficiency and power\n\nenergy change between two harmonic oscillators\n\ndifferential programming, RL-like scheme\n\nmissed the flawed thermodynamic definition\n\n### Clarkson University: reservoir computing\n\nreservoir computer: a kind of neural network, for forecasting dynamical systems but most of the parameters are chosen randomly. cheap but works well.\n\nESN Jaeger 2001\n\nclassic representation theorem by WOLD theorem\n\nworks because time soaks randomness. equivalent to logical NVAR (non-linear vector autoregression)\n\nenjoys universal approximation theorem, even linear with non-linear readout. \n\nVAR: vector autoregression\n\nVMA: vector moving average\n\nDMD\n\nmacket-glass equation as its example\n\n### Kim’s recurrent neural programming language\n\nalso reservoir computer. \n\nsilicon computers represent with binary and compute sequentially.\n\nneural computers use continuous symbols and compute distributedly\n\ndRAM: dynamical random access memory\n\nstore a reservoir into another reservoir\n\nanother hot topic is Koopman's theory\n\n### Learn complex fluids with rheo\n\nRhINNs (PhysicsINNs): inform the NN with underlying physics/rheology\n\nhis model is called RhIGNet, and an improved “multi-fidelity RHIGNets”\n\n### Data-driven selection of rheology-informed NN\n\nsame group as previous one.\n\nmodel works better as it goes more complex, while data wants model to be simpler.\n\nwe should prioritize data range over size\n\n### ML for Robot Locomotion in Flowable Materials\n\nrobot cars (RRP) climbing sand slopes\n\n### Predicting Clogging\n\nhopper flow\n\n### Analogical Reasoning to build Transferable Models\n\nmodel manifold: for a given data sampling method, points in parameter space can be mapped to a sampling space\n\nmanifold boundary approximation method\n\nsupremum principle\n\nWnt signalling network\n\n### [Evidence for Griffiths Phase Criticality in Residual Neural Networks](https://march.aps.org/sessions/M09/12)\n\nresidual networks’ has Griffith phase, due to the design (sth.  and renormalization)\n\n## N08: Systems Far from Equilibrium\n\n### Hopping Particles\n\nmath heavy\n\nDoi representation, Jarzynski relation, Doi-Peliti field theory\n\ngeneralized fluctuation-dissipition into non-equilibrium\n\nno coarse graining or slow modes\n\n### [A topological fluctuation theorem](https://march.aps.org/sessions/N08/2)\n\n$\\Delta S_M = \\frac{Q}{k_B T}n = \\gamma n$\n\nalso math heavy\n\ndetailed TFT: symmetry of particle in vortex field is protected by topology\n\n### TUR in Langevan processes\n\nTUR: thermodynamic uncertainty relation\n\n$\\frac{var(j)}j{}\\ge \\frac{2}{\u003c\\Sigma\u003e}$\n\nprobability distribution → rate function. don’t know how\n\n“caveat” seems an abused word, another is “ansatz”\n\n### reversal symmetry for cyclic paths from Notre Dame\n\ndetailed balance implies a reversal symmetry\n\nexpressed as Markov process: $\\frac{p_i}{p_j} = \\frac{l(i,j)}{l(j,i)}$\n\nhis work is just cycles in non-equilibrium systems\n\nfluctuation theorem for **cycle counts**. PRE 2021\n\n### emergence and breaking of duality symmetry in generalized thermodynamic relations from UNC chapel hill\n\nfoundation of thermodynamics, statistical mechanics, can be further founded upon classical or quantum mechanics\n\nthree times of Legendre transforms of internal energy ought to give 0, which breaks the reversibility of Legendre transform.\n\nT. Hill 1963: added some sub linear term\n\nthey used mean of finite measurements\n\ntheir understanding of thermodynamics is “nothing more than theories of probability”\n\n### Scaling of entropy under coarse graining\n\nentropy production rate\n\nin 1 dimension Derrida)\n\ndimension matters and data reduction may have non-trivial effects.\n\n### Kibble-Zurek effect, quench\n\nphase transitions\n\n### Discrete Non-Linear Schroedinger Equation by Federico\n\nphase transition again\n\n### anormalous thermal relaxation in unimolecular chemical reactions\n\nMpemba effect: hotter water freeze faster than cold water\n\n### [Perturbation spreading in a non-reciprocal classical isotropic magnet](https://march.aps.org/sessions/N08/10)\n\ntotally absent-minded\n\n### Stefan-Maxwell diffusivities of gas mixtures and liqiud electrolytes from Oxford\n\nelectrolytes, concentrated solution theory: Gibbs-Duhem relations\n\nvery math heavy, and words too small on slides. presentation visualization not so good.\n\nextended DFN model\n\nOnager’s regression\n\n## Q04: Non-equilibrium Thermodynamics: From Chemical Reaction Networks to Natural Selection II\n\n### Free energy transduction, chemical reaction network\n\nnon-linear reaction: at least second order reaction\n\nreactions → emergent cycles\n\n### Scaling relations of energy dissipation rate in non-quilibrium reaction systems\n\nstate space renormalization group\n\n### algebra on biological structures, Caltech and Cold Spring Harbor\n\nsecond quantization, field operators in multiparticle Fock space\n\nclassic particles in quantum language\n\nthe operation is the Wick contraction\n\n### decomposing, local arrow of time, interacting systems\n\n$\\dot I = D_{KL}[P(x\\rightarrow x')||P(x'\\rightarrow x)]$\n\n“entropy production” again: Prigogine 1947, Shcnakenberg 1976, Skinner Dunkel 2021\n\nobserve k elements in the system: interaction irreversibility $\\dot I^{(k)}_{int}=\\dot I^{(k)} - \\dot I^{(k-1)}$\n\n### information transmission by heterogeneous cell populations\n\nnoise can ruin cellular info sensing\n\nindividual cell senses better than population(?) **conditional mutual information** reflects this  \n\n### functional universality, microbial, thermodynamic constraints\n\nenergy-limited vs. nutrient limited\n\n### trade off:\n\nlaw of energy in biology; model organism being anaerobes B. theta\n\nPPi (pyrophorsphate) as energy source\n\n### 3D diffusion in E.coli\n\n## Physics of social interactions II\n\nfirefly, one research about vocabulary, one about spatiotemporal pattern\n\nbunblebees, SLEAP to identity action and track\n\nant, fire ant, BOBbots, multi-occupancy lattice gas, rule based model→observed result\n\nsocial polarization: vector force, but how agents are embedded into this space? no answer\n\nHKB system, third party induced bistability, symmetry breaking “HOW”!\n\n**Team formation**: [Modeling the Catalysis of Collaboration at In-Person and Virtual Conferences](https://march.aps.org/sessions/S03/9): Non-linear memory model, scialog dataset\n\nmirror game: no designated leader experts reach higher accuracy and velocities\n\nGranger causality analysis: not so advertised. experiment with flow switch. Presentation not so well given. \n\nBacteria reshape their surroundings to enable migration: use bioprinting. different pore size in media. non-motile vs. motile. motile E.coli can escape tight pores\n\n## **Learning dynamical models across physical systems**\n\n### **Learning dominant physical processes with data-driven balance models (walking)**\n\n- obtain kinetics without markers\n    - 2d median filter, viterbi filter\n    - **anipose**: combination of  2d filters, triangulation, minimize spatiotemporal variation in limb length\n- feedback model\n    - reflex model, central pattern generator\n    - tune the sensor neurons with laser\n    - proprioceptor model → state estimator → neural controller → (muscle model)\n\n### KNODES to learn complex dynamics and chaos by M. Ani Hsieh from UPenn\n\n- ocean data\n- neural ordinary differential equation (NODE).\n    - NeurIPS 6572-6583, 2018\n    - knowledge embedding enabled by NODEs. Knowledge-based learning of nonlinear dynamics and chaos. Chaos vol31. 2021\n    - applied to Hopf model, Stiff Can Der Pol model,\n    - inferring swarming dynamics from data\n    - deformable image registration\n    - model predictive control that controlled drones\n    - learning chaos\n\nscalar.seas.upenn.edu\n\n### physics of behaviors across scales\n\nlarge scale behavior (trajectory): variation explained by exploration-exploitation\n\nfine scale (posture): decompose into 5 primitive postures\n\nconnection: \n\n- transfer operators → slow emergent variables, subsume nonlinear to linear\n- use time series to max predict system to compensate missing d.o.f\n\n“hopping dynamics” again \n\nnon-ergodic drives that prevent the dynamics from relaxing to the steady state within measurement time\n\nthe landscape the worms seeking food is also time dependent\n\n### ML for biomechanics\n\nthis group’s interested is very scattered.\n\nfirst part just missed: learn macroscopic params in hydrodynamics as fields\n\nUnet to predict some cell protein density, identify relevant signals\n\nmodel: Oakes et al, BioPhys J 2014\n\nfruit fly tissue\n\nphysics-informed machine learning again\n\n### physics-informed machine learning: climate modeling and COVID 19 forecasting\n\nUnet also here. turbulent-flow net. basically a new network structure based on unet. \n\ngroup equivariance leads to another net called Scale Equ-ResNet in order to retrieve symmetry\n\nspatial-temporal neural p...: Baysian Active Learning\n\n## Collective Behavior in Biology\n\n### thermal TRP channels of vipers\n\n1K sensitivity of a channel; 1mK sensitivity of neurons\n\nindependent channels, over time of several independent measurements\n\nhypothesis: TRPA1 channels are embedded into a dynamical system near bifurcation. activated channels activates other channels \n\nquestion: does coupling break the independence of measurements? single channel info is ~1\n\n### growth difference in single and multicellular organisms\n\nexpanding epithelium,when the area grows linearly, the size only ~$\\sqrt{a}$, does this feedback to cells?\n\ncells divide at a smaller size\n\nlower bound of cellular size might come from genome size. CCND1labels DNA demage\n\nno mechanisms yet\n\n### collective signal oscillation\n\na lot of models\n\nkey features not heard clearly. adaptive spiking\n\n### cellular migration on substrate with topological defects\n\nthe way they defined topology number = -1 is not as my expectation\n\nFrancesca Serra group experiment\n\nnot too much work\n\n### synchrony and causality\n\nBo Sun again. granger causality again\n\nused a famous model not written down. Sounded like Nagumo model\n\n### living chiral crystals\n\nstarfish embryos\n\n### central nervous system\n\ngut motions and nerves in crafish\n\njust using 5-HT is not enough to compensate the cutting of nerve\n\n### role of position info in collcetive gradient sensing\n\nCramer-Rao bound by maximum likelihood estimation\n\nbenchmark was called tug-of-war model\n\ncontact inhibition of locomotion\n\n### self-tuned criticality amplifies signals in bacterial chemotaxis\n\nE.coli chemosensing: high cooperativity, perfect adaptation, large noise\n\nsymmetric spreading of CheA\n\n### division, migration, cell signaling\n\nMDCK monolayer\n\nvery biology heavy\n\na low fluctuation phase\n\n### criticality in Cochlea\n\nimpedance = pressure/height, WKB approximation\n\nhair cells are the sensors, undergo Hopf bifurcation\n\n### defining success for pairwise maximum entropy models\n\nmouse brain hippocampus, 2000 neurons\n\nin subgroups, swap 1/2 of the cells to minimize pairwise correlation coefficient\n\n## stoichastic thermodynamics and biological and artificial info processing\n\n### [Entropy generation during computation - is it really avoidable, even in principle?](https://march.aps.org/sessions/Y09/1)\n\n[Entropy generation during computation](https://march.aps.org/sessions/Y09/1): Szilard’s argument\n\nerase and copy operation of ribonsome works like a universal turing machine.\n\nin theory, logically irreversible process can be thermodynamically reversible, like **erasure**\n\n### statistic thermodynamics of communication channels\n\nMIT media lab\n\n$\\sigma(\\vec x_f)\\leftrightarrow DK(P(\\vec x_f)||P(\\vec x))$ KL Divergence\n\n2 models, 3 by 3 nodes with some lines connecting some pairs, didn’ get what they mean.\n\nthey call the thing NESS (non equilibrium steady state).\n\ntakehome: dissipative not monotuned increasing as info transmission increases. some channel more efficient than others\n\n### fluid intellengence: activities from learning and forgetting\n\ndiameter variable particles as intelligent agents\n\ndiameter function $D(\\vec p)$ symmetry determines symmetry of new distribution\n\npolicy optimization, reinforce learning to search behavior space\n\n### nonquilibrium dynamics of temporally responsive, single molecular automation\n\nsmart vs. inert: ability to perform temporal pattern recognition\n\na polymer that has multiple foldable positions. \n\ndual scale master equation\n\n### thermodynamics of biological signal propagation\n\ninformation: $lim_{T\\rightarrow \\infin}\\frac{1}{T}\\int MI(I,O)dt$\n\ndiffusion of signal,some equations, etc, parameters include size of source and receiver\n\nimplications: \n\n- cells don’t talk because they are too small\n- E.coli phosphotases are close to ion channels (missed, not sure)\n\n### dissipation-accuracy-speed tradeoffs in computation on-;attice self-assembly\n\nLandauer principle\n\nD. Woods. Nature 2019\n\ndiamond shape 2-in2-out tiles that \n\n### Bayesian mechanics for interacting systems\n\nLangevin equation\n\nNESS again\n\n### wipe a bit with no energy cost by pypassing Liouville theorem\n\nLiouville says Halmitonian system must be incompressible in phase space\n\n$H_{ers} = H_0+gH_{contraction}$\n\nmicro-canonical energy shell\n\n### **Stochastic Thermodynamics of Finite Automata**\n\nnot following, deterministic finite automata\n\n### **Nonequilibrium thermodynamics of uncertain stochastic processes**\n\nwanted to follow but failed\n\n### glycan, function of Golgi constrain its morphology\n\nalso used KL divergence here.\n\nlooks like there has been no experiment, maybe I missed it. \n\n### **Stochastic thermodynamics of anomalous diffusion generated by scaled and fractional Brownian motions**\n\nStochastic thermodynamics: ST\n\nfractional Brownian motion: fractional gaussian noise (no idea what)\n\nfluctuation-dissipation is broken by memory in noise\n\n### **Optimality in biological proofreading**\n\nDNA replication error rate 10^-8 ~10^-10\n\nexperiment K_D/K_C = 10^-2, observed error rate 10^-3~10^-4, discrepancy\n\nPareto optimal fronts of kinetic proofreading\n\n**generalized Hopfield model**\n\nspeed-dissipation trade off again\n\n## Dynamics in Evolution\n\n### Multicellular\n\nSize→nutrition gradient→differentiation \n\nMulticellular yeast! T yeast\n\nBozdag\n\nNon-reformable→mechanical challenge \n\nThe aspect ratio increases significantly after the size grows\n\nAggregation not happening; Tangling spans the bulk\n"},{"slug":"notes-during-APS-march-meeting-2023-Las-Vegas","filename":"2023-03-12-notes-during-APS-march-meeting-2023-Las-Vegas.md","date":"2023-03-12","title":".tex | Notes during APS March Meeting 2023 in Las Vegas","layout":"post","keywords":["tex","phy"],"excerpt":"Notes during APS March Meeting 2023 in Las Vegas","content":"\nSome interesting sessions:\n\n![5FB0A373-E7B7-45FF-97A5-81F9392CBB02.jpeg](/blog/photos/2023-03-12-5FB0A373-E7B7-45FF-97A5-81F9392CBB02.jpeg)\n\n![E35F718E-38ED-432D-AC00-44E8D8893CC6.jpeg](/blog/photos/2023-03-12-E35F718E-38ED-432D-AC00-44E8D8893CC6.jpeg)\n\n![A44D0789-44CC-42F5-85D1-E0CA93952712.jpeg](/blog/photos/2023-03-12-A44D0789-44CC-42F5-85D1-E0CA93952712.jpeg)\n\n![7E0D2844-3BC8-4BEF-BABA-2818EEBA2C15.jpeg](/blog/photos/2023-03-12-7E0D2844-3BC8-4BEF-BABA-2818EEBA2C15.jpeg)\n\n![A727558B-D110-48A1-B125-67A7067BFBD5.jpeg](/blog/photos/2023-03-12-A727558B-D110-48A1-B125-67A7067BFBD5.jpeg)\n\n![B635604C-B9BD-4A34-9767-AD7338A80765.jpeg](/blog/photos/2023-03-12-B635604C-B9BD-4A34-9767-AD7338A80765.jpeg)\n\n## Network theory and application to complex systems\n\n1st talk: **Avalanches Dynamics** and its critical points, cortes dynamics, very long, missed first ~10 minutes, hard to follow, coarse grained 朗道-金斯堡 theory\n\nFokker-Planck equation, Fourier expansions, use as parameters of ____, choose a closure to solve\n\n## Machine learning in complex systems\n\nFractional nn for continuous modeling of complex fluids: Caputo method for fractional derivative \n\nInferring forces in many-particle systems using physics tailored ml: Chinese student with Ilya from Emory. Bruckner 2021 PNAS, cell. His work, dusty plasma, non reciprocal. Mutual Interactions, with environment, linear damping. not clear details of the nn \n\nEcho state network: Arizona state, Chinese student. A branch of reservoir learning. Switching states not always successful. Bad presentation. Don’t even know what question he is learning with this nn. Could be Lorenz system\n\nHopper Flow: particles passing a narrow gate in 2d. Classify problem: flow, clogging, clogged, empty arch. Direct interpretability: linear SVM. \n\nCommitter of Reactive Events in and out of equilibrium: UCB. Variational methods for time-dependent backward Kolmogorov equation. decompose rate into contributions from different degrees of freedom. \n\nPredict Microfluid properties with ML: oil droplets into water, predict the droplet size. Monodispensed phse, poly  (jetting). Pretty small network and small dataset. nice GUI\n\nML for inverse design of metameterial. Combinatory/Discrete, imbalance. building block → unit cell → metamaterial. Ogliomodes\n\nLong-term behavior with liearn regression. Moody 1944 moody diagram. Studied Lorenz system. Use short trajetories to predict long term properties. Period orbit theory \n\n## Quantitative cell physiology I - shape and size\n\nmissed 1st talk for breakfast.\n\n### Beyond G1/S regulation: how cell size homeostasis is tightly controlled throughout the cell cycle?\n\nG1/S not the main control cell size variation; CV reduces across cell cycle\n\nrate positively correlated to size, but not exponential. this should lead to different mass density, but it is maintained tightly stable.\n\n### Cell cycle dependent growth regulation in bacteria.\n\ndata  show that E. Coli cell growth is super-exponential.\n\nmodel: ribonsome growth is exponential, \n\nB. Subtilis growth rate is not monotonic\n\nScott, M, 2014 Mol Sys Bio\n\ntome averaged mass function and allocation function are equivalent.\n\nA model, but didn’t see how it’s correlated to cell-cycle\n\n### E. Coli coordination of cell dimensions to preserve macromolecular densities between compartments\n\nshape is also important, not just dimensions. \n\nperiplasm is the region between inner and outer membrane of E. Coli. Its colume fraction scales differently with the width and length of the cell\n\nmeasure periplasm density by shocking and destroying the outer membrane and measure protein amount\n\ncell aspect ratio is relatively controlled by changing periplasm biomass fraction and average cell width. Testified in multiple mutants\n\nGrowth rate is not a control parameter\n\n### Molecular noise on cellular growth\n\nstandard model of size homeostasis: timer, adder, sizer\n\nconsider noise, get lower bound of size \n\n### Dynamic proteome trade-offs regulate cell size and growth in fluctuating nutrient conditions\n\ngrowth rate and size control; in fluctuating nutritions, there is overshoot \n\nassums: cell maximize flux; division protein co-regulated with metabolic sector\n\nmodel predicts cells prioritize division over biomass accumulation durign downshift\n\n### Simulation about mechanical properties of Peptidoglycan (PG)\n\nUniversity of Jena, Balance of the Microverse. Very technical, very detailed\n\n### Teichoic acids as organizing center of growth in B. Subtilis\n\ninhibiting its synthesis immediately loses the cells’ rod shape; fluorescent images are uniform across the whole cell.\n\nperturbing synthesis causes growth rate shutdown after loss of rod shape\n\nPonA mediated crosslinking is required for growth in the absence of wall teichoic acids\n\nproposed a model, not quantitative, seems not testified.\n\n### coarse grained cell wall simulation\n\nturgor pressure: 膨胀压\n\nanisotropic deformability\n\nsurface evolver\n\n### Nanostrctural changes of Peptidoglycan\n\nPenicillin binds enzymes with the last step of PG synthesis.\n\ntreat with mecillinam; use AFM, changes local molecule orientation \u0026 PG organization\n\ngaps in the PG network, not orientation, caused the death\n\n### Robustness and Fragility in Yeast\n\nUmap for metabolome data\n\n### How Type 1 Fimbriae Help E. coli Adhere to Interfaces?\n\nWhat is a fimbriae in bacteria?\n\nThin,  hairlike appendages, 1 to 20 microns in length and often occurring in large numbers, present on the cells of gram-negative bacteria, particularly Enterobacteriaceae and Neisseria.\n\nthey contruct mutants and can both adhere and swim.\n\n### ZnO kills bacteria\n\nHypothezes: metal ion production, generation of reactive oxide \n\n## March 7, 2023 Ecological Dynamics\n\n**Spatial drug heterogeneity:** a row of multiple wells, some of them have drugs. Mean field survival criterion: **$$L=\\pi\\sqrt{}$$**, and beyond mean field (some purtubation theory borrowed from quantum to reaction diffusion equation).\n\n### Evo adaptation on cell shape by Katie\n\nmechanical factors give same-fitness cells different growth rates in a colony. cells labels with 2 colors of fluorescence and grow in a mixed colony\n\nmechanical forces slows down natural selection, both directions\n\nused yeasts and grow for 10 days. select edge and center and continue growing. center still similar, but edge cells become elongated. “The elongation is a mechanism to get out of nitrogen starvation”(?)\n\nsequenced ancester, center, and edge cells. Found 9 genes\n\n\u003chr class=\"slender\"\u003e\n\n**The fate of beneficial and deleterious mutations in range expansions driven by long-range dispersal:** luckiest better than fittest, spatial factors dominate.\n\n**Speciation at an expanding front**: superdiffusion → stepping stone model. Want to explain how homogeneous population evolve into species→ fitness model \n\n### F13: Emerging Principles and New Developments in Non-Equilibrium Thermodynamics, Natural Selection and Chemical Reaction Networks - Nonequilibrium thermodynamics of chemical reaction networks\n\nCRN: chemical reaction networks, and this is about \n\ncircuit theory for CRN: arxiv-2210.08035\n\n- modules do not share internal species\n- emergent cycles and effective reactions\n\n## 203 cell model\n\n### Active transport and Cell self organization\n\nA lot of microtubules \n\nCCB-FL\n\n![214D3DB8-7A99-42AC-8C18-051171564D6D.jpeg](/blog/photos/2023-03-12-214D3DB8-7A99-42AC-8C18-051171564D6D.jpeg)\n\nSteam flows in large cells B. Corti 1741, fruit fly oocyte \n\nFlows in organelle positioning. Zidovska PNAS 2013, 2018. Gatlin lab U Wy \n\nMicroscopic mechanisms about micro tubules is ill understood \n\naLENS a living ensemble simulator \n\n1model: Cytoplasm as confined Newtonian fluid\n\nAnother model: coarse grained stokes force: fruit fly oogenesis \n\n![C622BD7E-E638-4194-A563-096F73209132.jpeg](/blog/photos/2023-03-12-C622BD7E-E638-4194-A563-096F73209132.jpeg)\n\nskellySim\n\n### Di Pierro: physical genomics\n\nNucleosome\n\nbut his title is **Landscapes of Genomic Architecture Across Evolution**, which is not \n\nideal chromsome: every time  2 loci at genomic distance d come into contact there is a gain/loss of free energy $$\\gamma(d)$$\n\n### Stochastic model for cell survival and evolution\n\nSingle threshold and double threshold \n\n## metabolism and growth\n\n### Adaptability and sensitivities in gene regulation out of equilibrium\n\nSpare able energy \n\nMatrix tree theorem: inter alios\n\n### biofilm electric shocked\n\nBacteria use membrane potential to regulate bio processes: cell division, antibiotic resistance, metabolism \n\nElectrical stimulation \n\nBiofilm: B subtilis: motile cells and matrix forming cells\n\nAccumulation of motile cells around electrodes \n\nProliferation or switch? Proliferate \n\n### Self organization waves on huge single cell living on light\n\nexternal driving force main  reason?\n\nAlso internal limit cycle?\n\n### CRISPER array for E. coli under antibiotic exposure\n\ncrisper cas 12-a\n\nBeta Score , different in different media\n\n### ZnO Nano particles on bacteria in magnetic field\n\nGram staining \n\nSynopsys quantumatk\n\n## ecological dynamics\n\n### design microbial community\n\nGoal: metabolite production \u0026 degradation; pathogen inhibit \n\nEmerge, \n\nfunction as a design problem = f(composition)\n\nTrivial idea: open the hood and see the interaction \n\nAnother way: borrow from molecular level: how sequences gives function \n\nStatistical design: blue and red fluorescent proteins, separated by 13 amino acids, fit color as 13d binary\n\nSuccess because landscape simple \n\nEcology: if one can predict function from presence absence of a strain\n\nAbsence =-1 not 0, for similarity with Frourier expansion \n\n500 communities of 27 strains, predict production of a chemical \n\nsampling a very small space but result good, (predict/true value diagram)\n\nAlso predicted biomass (OD)\n\nAdditive property is the key to success \n\nCan get explained variance by the order of \n\nRuggedness/slope\n\nCompare with Generalized Lotka-Volterra\n\nMetabolic cascades makes landscape rugged.\n\n### connecting global patterns to ecological processes\n\nGlobal soil samples\n\nEnvironmental filtering cannot explain some of their data\n\nEmerges from interactions \n\n### Mapping prediction error versus information content of coarse-grained descriptions of microbial ecosystems\n\nJacob moran\n\nFramework to compare different coarse grained descriptions: partitioning, error-information diagram\n\nTo distinguish mechanisms of emergent simplicity: diversity in the partition \n\n## Materials that work by themselves\n\n### biotic and abiotic\n\nAdaptive d spring\n\n## Pattern Formation\n\nmissed 1st talk of 30 min\n\n### Morphoelastic growth to model aortic geometry\n\naortic 主动脉\n\nintegrated gaussian curvature\n\nAI/ML needs a lot of data, so they need explicit models\n\nNHLBI: National Heart, Lung, and Blood Institute\n\n### experimental on chaos in cardiac tissue\n\nkept in Tyrode solution\n\naction potential duration + diastolic interval = period\n\nTISEAN (a [Nonlinear Time Series Analysis](https://www.pks.mpg.de/tisean/) tool) to get Lyapunov exponent\n\n### excitability at interface\n\n2 bio-electric tissues, one only sodium channels, the other only potasium channels. so no spikes on their own. then shed laser on them.\n\nA as fast variable, B as slow variable. A is x^3 shaped, whie B is -x shaped, then \n\ncan also exlpain BZ reaction with some sub-reactions separated, and lumberjack and forect and desert\n\n### light induced cortical excitability\n\nshape control strategy in meiotic starfish oocytes.\n\nrho protein. part of some signaling pathway, kymographs\n\nseems to be my competitor on postdoc project in MPI-PKS\n\npole to pole opto-GEF accumulation aymmetry\n\npresentation overtimed, not well delievered\n\n\u003chr class=\"slender\"\u003e\n\n**Exploring regular, exotic and turbulent flow states in active fluids via the framework of Exact Coherent Structures**: too fast, totally lost. Who tm knows what is shadowing\n\n### Pattern plant roots\n\ntropism 向性: gravitropism, phototropism\n\nroots on obstacles: skewing, coiling, waving\n\nnumerical scheme: separation of timescles, derivatives of intrinsic curvature over time, against cross section growth rate gradient.\n\ndecoupling the tropic dynamics parallel and perpendicular to the slope.\n\neach curing given an energy, can predict transition between states\n\nskewing is different, seen in mutants. It’s the intrinsic twist \n\n### cdc42 mobility changes yeast shape and survival\n\nAs title says \n\n### Jagged ligand improves notch signaling\n\nNotch, delta, jagged \n\n## Synthetic biology\n\n### stochastic model SELEX\n\nto select aptamers with highest affinity : [https://www.basepairbio.com/what-is-an-aptamer/](https://www.basepairbio.com/what-is-an-aptamer/)\n\nOptimal policy: A1 A2\u003e\u003eT, T\u003e\u003e1\n\n### Cytoplasm is organized, crowding\n\nCrowded with macromolecules\n\nCrowding of cell free gene expression change noise; simulations \n\nCrowding gives attraction to membrane \n\nSimulations \n\nImpacts transcription and translation in cell-sized vesicles\n\n### Elisa from Cornell : titrable in vivo Cre-loxP in E. coli\n\nSomatic differentiation \n\nEr binding site, beta estradiol\n\nDH10B RepOrter yfp \n\n## Research in Germany\n\n### MPI\n\nWrite emails with info from website \n\nhumburg scholarship\n\n6+2 years super generous funding enough to hire 2 postdoctoral, to early scholars\n\nForeigners 9 years after doctor degree has to be employed, so many leaves\n\nUniversity and Scores important, papers slow so not so important\n\nDPG MEETING\n\n140 directors, each has a way of sponsoring \n\nAndrew Mackenzie \n\nGo for Nobel prize, not professorship \n\nWant to hire people with 20 yers of career left\n\nSix months danger period, then 3 years safe\n\n### Cell fate transition\n\n### cell type order parameters\n\nSaw last year \n\nOrder parameters: \n\nscTop python package\n\n### Class IV da neurons Laval development\n\nMacroscopic measurable: dendrite Densities, front velocity \n\nDendrite grow in stochastic manner \n\nMean field descriptions: grow, shrinking, stall states\n\nmaster equations \n\nVery in detail\n\nYale integrated graduate program in physical and engineering biology \n\n### synNotch expression in fruit fly\n\nDelta notch: contact based intercellular signaling \n\nVortex based model framework \n\n## Intracellular transport and organization\n\n### super resolution\n\nSTORM, PALM, FPALM, all in 2006\n\nCorrelative live cell and super resolution microscopy\n\nPost translation modification: Tyrosinated/detyrosinated, de- is important , sparse and stochastic \n\nDe- VASH interact with MT \n\nMean square displacement analysis \n\nSimulation: \n\n### In vitro experiment of Microtubules\n\n## switch and detach at intersections of microtubules\n\ndiaplacement≤path length, from experiment \n\n### Kinesin-1,2,3, bidirectional transport\n\nDifferent detect rate\n\nKinesin compete for dynesin in a tug of war\n\nsegment analysis of slow trace \n\n### motor\n\nMTOC, convergence and target scretion of granules\n\nNatural killer cells\n\n### Growth of axonal caliber\n\nAxonal neurofilament as caliber modulators \n\nSix state model Jung and Brown 2009, Li etc 2014\n\n### intracellular drug transport in single cell\n\nHoechst dye to measure effenciency\n\n### alien proteomics, diffusion through nuclear pore\n\nNuclear pore complex\n\nFG-nucleoporin (IDR)\n\nActive transport vs passive diffusion \n\n\\Delta F(R)\n\nInject frog oocyte into E. coli \n\n### cone shaped hiv capsid through nuclear pore complex\n\nFree energy of nuclear protein \n\nNarrow end enter first\n\n### ER for transport Calcium ion\n\nCa bound to high capacity buffer proteins in ER\n\nlocally release in spikes\n\nReaction-diffusion model for calcium and buffer proteins\n\nlocal cytoplasm concentration\n\nin ER Lumen: proteins spread super diffusively; active runs along tubules, from single particle trajectories\n\naltering morphology by morphogen expression: Rtn3 overexpression; Atlastin knockout\n\n## collective behavior\n\n### biofilm=cells+matrix\n\nSPP2389- emergent function of bacterial multicellularity\n\nNg bacteria \n\n### flow-dirven cancer cell migration\n\nautologous chemotaxis is a mechanis cancer cells fall onto XXX\n\nchemotaxis vs. mechanosensing\n\nchemotaxis fails at high cell density\n\n### motility induced phase separation of deformable cells\n\n所谓 phase separation，就是一筐红绿豆自己分成两堆\n\nrigid particles aggregate and sort without adhesion\n\n2d layer of deformable active brownian particles\n\nsimulation: high deformability easier to phase separation, shorter collision time\n\nBioPACIFIC program at UCSB and UCLA\n\n### Geometric programming of collective sensing of neural cells\n\nOregon State University, Chinese speaker\n\na cross shaped titch, with five wells at 4 tips and 1 crossing, (infinitely peridic? not sure)\n\nflow switch of GM and ATP\n\nlonger period leads to more well synchronized nodes. edge probability pattern\n\n### Ear, hair cells tune cochlea to criticality\n\nimpedance\n\nseems have heard last year\n\nBo Sun Gourp member?\n\n## Social Interaction\n\npretty big hall actually\n\n### The Use of Laser Speckle Spectroscopy for Subtle Detection of Infantile Ants\n\nGeogeria tech\n\npixel photon intensity, not direct movies, to detect movements. also sounds like something last year.\n\n### density functional fluctuation theory to model basketball\n\nNBA has position data every 40 ms!\n\n50 feet x 50 feet heatmap of shots\n\npostulates: location of ball changes player spatial tendencies, 2 types of player defense and offense \n\n![8795A615-E1CB-44C4-B02B-3AB5884AA384.jpeg](/blog/photos/2023-03-12-8795A615-E1CB-44C4-B02B-3AB5884AA384.jpeg)\n\nforecast a play, evaluate a position, improving a player’s positioning.\n\n### honeycomb formation under geometric frustrations\n\ndecentralized division of labour, to build honeycomb\n\na lot of energy cost, not always on flat plane. \n\n3d printing to prepare substrate for bees, for data repeatability. basically they make plastic cells of different sizes, and see how bees build upon them.\n\ngeometric frustrations: shifts, tilts, sizes\n\nforcing smaller cell size → X-ray microscopy\n\nsome ~24% of the cells are filled with wax\n\nlarger cells → tilting of wax walls, can be 3x bees’ normal honeycomb cell size\n\n### winner and loser emergence in zebrafish fights\n\nwhat is a fight, what is a winner and dominance, how fights start and end.\n\n6 pairs fo fish, 3 body points of 2 fish in 3d, experimental tracking\n\ndefining winner: post-fight position asymmetry\n\nfight-bout micostructure to **mesoscopic** fight-maneuvers. \n\nphase space: $$(\\theta_W, \\theta_L, D_{pp})$$\n\n### Chimera states among synchronous fireflies\n\nLewis and Cratsley Annu. Rev. Entomol. 2008. temporal patterns of different species\n\n[https://fireflyexperience.org](https://fireflyexperience.org) \n\n$$\\dot\\theta_i=\\omega_i+\\frac{K}{N}\\sum_j\\sin(\\theta_j-\\theta_i)$$\n\nchimera states for couples oscillators: part of the agents are synchronized, some are not, they coexist\n\nexample: unihemisphere sleep in marine mammals\n\nSouth carolina, Congaree National Park\n\n### Synchronization of human networks\n\nMoti Fridman (Bar Ilan University)\n\n![3790D8EF-864E-4107-9C44-2CD22D76AF8F.jpeg](/blog/photos/2023-03-12-3790D8EF-864E-4107-9C44-2CD22D76AF8F.jpeg)\n\nprevious works deal with the case where everyone sees everyone\n\nviolin players，但是成员互相看不到对方，one control position(? why need it)\n\nbidirectional and delay \n\neven number of plays give checker board out-of-phase pattern \n\nodd number gives frustration, they need to ignore one of the plays, changing the connectivity of network, glass state\n\nunidirectional network: humans change coupling strength at will when there is no delay\n\n## Mechanics of cells and tissues\n\n### division and differential adhesion drive cell self-segregation during embryo morphogenesis\n\nVoronoi cell model for cell monolayers\n\ncontact tension controls tissue structure\n\n### mechanosensing breast cancer spheroids\n\nanother Bo Sun group member.\n\nto segment and track the cells in 3d: PEN projection enhanced network, on top of a Cellpose 2d neural network.\n\nclassify cell shape phenotypes with SVM\n\nfilopodia leads invasion\n\n### kinetics of cell aggregation\n\ncell-cell contacts: Guo et al Biophysics J. 2006\n\ncell-substrate interactions: \n\n感觉是个大杂烩，啥都有\n\nmatrix mediated mechanical interactions, so what does matrix mean?\n\n\u003e ChatGPT:\n\u003e \n\u003e In biology, a matrix refers to a non-living or extracellular substance that provides structural support or serves as a medium for cellular interactions. The matrix can be made up of various components such as proteins, polysaccharides, and minerals, and it can be found in different biological contexts, such as bone, cartilage, connective tissue, and the extracellular matrix surrounding cells.\n\u003e \n\u003e For example, in bone tissue, the extracellular matrix provides support and rigidity to the bone and is primarily composed of collagen fibers and mineralized calcium phosphate. In cartilage, the matrix is composed of collagen and proteoglycans and serves as a cushion between bones in joints. In connective tissue, the matrix provides support and allows for the exchange of nutrients and waste products between cells.\n\u003e \n\u003e In summary, the matrix is an essential component of many biological tissues and provides structural support and a medium for cellular interactions.\n\u003e \n\n### epithelia tissues as hexnematic liquid cystals, and instabilities\n\nHeager et al,  J. Exp Med 2020\n\nhexatics in a channel, momentum conservation and orientation dynamics\n\ndispersion relation and unstabilities\n\nvery classical dynamics study, as classical as those on textbooks\n\n### Multiscale measurements of mechanical stress in 3D co-cultures using a deformable micro-device\n\n巴斯德研究所\n\ncell level reponse to tissue level mechanical xxx\n\nmechanic stress on each cell depends on position and neighbors; global deformation is accumulation of local stressees.\n\n### Mechanoregulation of biofilm homeostasis\n\nEPFL, Lausanne, Switzerland \n\nbiofilm in hydrogels\n\nsize control hypotheses: mechanical, growth rate, cell death\n\n### mechanosensing of embedded spheroids and cell invasion\n\ninserted in a fibrous network consisting of collagen, and other proteins\n\n3d **vertex models**\n\n[https://github.com/ZhangTao-SJTU/tvm](https://github.com/ZhangTao-SJTU/tvm) \n\n### density variation of fibroblasts near topological defects\n\nridges to introduce topological defects, ridge height as proxy for ridge strength\n\nshape dependent rate of division as the mechanism\n\ncell density changes\n\n### slicing softly with paper\n\npaper cutting machine to use paper cutting into gels\n\nsmall angle, thickness: buckling → cutting → indention\n\nlarger angle, only buckling → indention\n\nappled stress must exceed critical threshold, but below buckling threshold\n\ndangerous thickness: 65 microns\n\n### Vasculogenesis\n\ncardiovascular system is 1st organ system to form in a vertebrate embryo. 2 steps: vasculogenesis and angiogenesis\n\nthen totally lost\n\n### Cell shape and fate changes in 3D model of mammalian hair follicle development\n\n3d vertex model\n\n2 hypotheses: 1. changing cell autonomous properties; 2. extrinsic force. model 2 wins\n\nWTH is E15.5 morphology\n\nLisa Manning group\n\n### Emergent chirality in active solid rotation of pancreas spheres\n\n\u003e **ChatPDF and Humata.ai**\n\u003e\n\u003e ChatPDF is almost unusable, there is no context, and cannot break lines. They even answered in French first\n\u003e \n\u003e\n\u003e Humata.ai: The paper investigates the rotational dynamics of spherical tissues using murine pancreas-derived organoids as a model system. The study uses an active **vertex model** to demonstrate how the interplay between traction force and polarity alignment can account for distinct rotational dynamics. The study also investigates the emergence of chiral symmetry breaking in multicellular systems using experiments and simulations of rotating pancreas spheres. The authors propose that topological defects in the polarity field underlie the symmetry breaking. The article discusses the role of geometry in the behavior and properties of epithelial tissues, including cell division, patterning, and left-right symmetry breaking.\n\u003e \n\n语速还挺快\n\ngeometric constrains \n\n3D上皮 collective dynamics, mouse pancres self organize into a sphere, inside filled with lumen\n\ndistinct rotational modes: rotation, rotation arrest, drift in rotation axis → mechanism?\n\nactive certex model: force balance, polarity dynamics\n\nsolid rotation, or fluid exchange of neighbors. solid body rotation fraction.\n\nrotation axis drift is reminiscent of yielding like transition\n\ncell shape orientation is almost stable, not a uniform random distribution when there is rotation, but uniform random without rotation\n\npolarity dynamics: 2 types of solutions\n\nother structures: cubodal, compact, multi-lumen: **this is the proposed future direction**\n\n## collective behavior\n\n### spying on bacteria in 3d\n\nIsolated cells:\n\nBall pit for E. coli cells for them to tumble \n\nActive hopping and trapping\n\n\u003chr class=\"slender\"\u003e\n\ngranular hydrogel matrix supports 3d printed colonies \n\nCollective motion, not diffusion\n\nConsumption and chemotaxis \n\nHow to keep front together: Change initial position. Chemostatic velocity\n\nSensing = F(c)\n\n### multicelluarity w/ Differentiable Molecular Dynamics Simulations\n\nJax-MD: [github.com/google/jax-md](http://github.com/google/jax-md) \n\nmachanism for regulating division\n\n“learned params consistently yield homogeneous growth”\n\ndon’t like this project very much. which part is done by the speaker?\n\n### collective behavior of swimmers in intermediate Reynold number\n\nmodel of dumbbell in fluid, 2D [torque and force] free\n\npair of dumbbells: puller and pusher\n\n### Chemotactic motility-induced phase separation\n\n**Chemotactic**\n\nstudent of the invited speaker of this session, award winner.\n\nmotility-induced phase separation: 抖豆子抖成堆\n\nchemotaxis: directed motion in response to chemical gradient\n\nparticle volume fraction, chemoattractant concentration, \n\nsets up an allowed region of phase separation\n\narxiv:2301.12345\n\n### Anomalous collective dynamics of dividing chemotactic agents\n\nChemotaxis + population dynamics\n\nnon-equilibrium continuous phase transition\n\nsome kind of reaction diffusion equations for $$\\rho, c$$ \n\n**chemotactic** interactions can alter {} near critical point\n\n### Crowds with large pressures\n\nnon-equilibrium \n\nmicroscopic: Vicsek model, Helbing-Molnar model; macroscopic: mechanical spectroscopy of insect swarms, equation of state\n\nsys behave like equilibrium: Silverberg 2013 PRL\n\nwhy works? difference of time scales: relaxation time \u003e\u003e collision time.\n\nagents packed: potential energt not equi-partition, but kinetic energy are.\n\n### phenotype probability distinguishes mutants from wildtype social bacteria\n\nM. xanthus form persistent or not persistent dunes.\n\n## Morphogenesis\n\n### tension triangle\n\n8:30-8:40\n\nOrder and anisotropic form coherent T1 that drive flow\n\nA lot of UCSB, actually one group having 2 talks back to back \n\nTension isogonal decomposition \n\nCould be applied to wing disk, neurolation, \n\n### from gene to geometry\n\nCells → sheet → tube → cortex \n\nmulti view confocal, Bayesian based deconvolution, deep tissue imaging \n\nKnockout and shape differs\n\nTubLUAR: tube like surface Lagrangian analysis resource\n\nHow gene → calcium signaling → muscle contraction → endodem strains → organ shape\n\n## plant Biology\n\n### root hair\n\nself anchoring, 仿生根 robot, can dig into fake moon soil, for soil sampling and sensor implanting\n\n### Heterogeneous cellular growth in root\n\nDirectly connected to organismal shape\n\ncell behavior to organisms \n\nDiscrete element simulation \n\nMeristematic and elongation zone\n\n### Turing pattern in vegetation\n\nExperiment: Australien fairy circles regenerated\n\nWebGL and JavaScript\n\n### network architecture of plant pathogen interactions\n\n![D185C1B7-F5EB-4629-B767-0ECF7F72C591.jpeg](/blog/photos/2023-03-12-D185C1B7-F5EB-4629-B767-0ECF7F72C591.jpeg)\n\n![03A05C17-8E19-4B30-B449-1A99D02093A9.jpeg](/blog/photos/2023-03-12-03A05C17-8E19-4B30-B449-1A99D02093A9.jpeg)\n\nTranslation/degradation, xxx, xxx\n\nSimple model. complex systems \n\nRockefeller U.\n\n### water solubility of hydrophobic xxx with phytoglycogen(PG) nano particles\n\nUniversity of Guelph, in Canada\n\nAstaxanthin-\n\nCarotenoid solubilization strategy\n\nOSA modification changes absorption spectra \n\n## Emergent behavior in biological systems\n\n### Emergent Spatiotemporal Communication Patterns in Insect Swarms\n\nChemical signaling (bees) and visual signaling (fireflies)\n\nWorker bees scent to amplify the feromon of the queen to locate the queen \n\nAlso Happens in fed bees to locate food\n\n\u003chr class=\"slender\"\u003e\n\nVisual signals by fireflies \n\n非线性动力学的课本居然有 synchronization, slide 和之前看过的一样\n\nCan only see a small portions of fireflies in the wild\n\nControlled environment \n\nTalking to fireflies with LED\n\n### anatomy of E. coli colony expansion\n\nHeight depends on glucose concentration \n\n### mechanical information processing by action cytoskeleton\n\nFocal adhesion\n\nPretty biology heavy\n\n### CryoET\n\nTomography \n\nCapture topological features on sparse density data\n\nImage voxels and compare with alpha fold candidates.\n\nIs this risky for cyclic reasoning?\n\n### protein assembly\n\n最后一天，没有心气记笔记了……\n"},{"slug":"python-subprocess-run-bash","filename":"2022-07-08-python-subprocess-run-bash.md","date":"2022-07-08","title":".py | Executing Bash Commands with python.subprocess","layout":"post","keywords":["md","py"],"excerpt":"Translated by ChatGPT.","content":"\nThe touchscreen on my laptop has a crack from a drop. At first, it wasn’t a problem, but in recent weeks, it has worsened, and sometimes the cursor goes wild and doesn’t respond to commands. So, I needed to disable the touchscreen as an input device.\n\nIn the Xorg desktop environment, you can use `xinput list` to display all input devices along with their corresponding ID numbers. Once you locate the device’s ID, you can disable it with `xinput disable ##`. Generally, the ID number is stable, so I directly added the disable command to `~/.bashrc`.\n\nHowever, I recently took out my long-neglected Raspberry Pi and, for fun, bought a compact 60% mechanical keyboard. After plugging in the keyboard and logging in with my password, guess what? The new keyboard stopped working, giving me a real scare.\n\nSo, now I need to check the input device IDs each time and pull out the IDs related to the touchscreen (there's more than one) from the `xinput list` output. I need to assign them to a variable and then input this variable into `xinput disable #`. This sequence of actions is beyond my Bash skills, so I considered using Python for the intermediate steps: that is, getting Python to read the Bash command results and write the Bash command statements.\n\nA Google search brought up this result: [https://stackoverflow.com/questions/4256107/running-bash-commands-in-python](https://stackoverflow.com/questions/4256107/running-bash-commands-in-python). After a quick look at the official `subprocess` documentation, I wrote the following snippet, saved it as `~/disable_touchscreen.py`, and added `python ~/disable_touchscreen.py` to my `~/.bashrc`.\n\n```python\nimport subprocess\ncheck = subprocess.run([\"xinput\",\"list\"],capture_output=True)\nfor line in check.stdout.decode(\"utf-8\").split(\"\\n\"):\n    if \"touchscreen\" in line:\n        device = line.partition(\"id=\")[2].partition(\"\\t\")[0]\n        disable = subprocess.run([\"xinput\",\"disable\",str(device)])\n        if disable.returncode == 0:\n            print(f\"Successfully disabled touchscreen device {device}\")\n        else:\n            print(f\"Failed to disable touchscreen device {device}\")\n```\n\n- The core of this code is `subprocess.run()`. The first parameter is the command passed to Bash, which is a list where each element corresponds to each part of the Bash command split by spaces. To get the command’s output, you need to add the `capture_output=True` parameter.\n- The return value of the function is a special data structure. If the command executes successfully, the result is in `.stdout`, which is a binary string `b'xxx...'`, so it needs to be converted to a string with `.decode(\"utf-8\")`.\n\nIf only using Bash:\n\n- My Bash skills can get as far as `xinput list | grep \"touchscreen\"`. Here, `|` is a pipe, which passes the output of the preceding command as input to the following command. To use a pipe in Python, you can refer to this question: https://stackoverflow.com/questions/13332268/how-to-use-subprocess-command-with-pipes.\n- The result of the pipe still needs to be trimmed to get the ID, i.e., `line.partition(\"id=\")[2].partition(\"\\t\")[0]`. This would probably require `awk`, which is challenging but learnable, and at least I know what to learn.\n- But assigning the ID to a variable and then passing this variable into `xinput disable`—I don’t even know where to start with that.\n"},{"slug":"python-import-script-module-package","filename":"2021-12-11-python-import-script-module-package.md","date":"2021-12-11","title":".py | Import Existing Code","layout":"post","keywords":["md","py"],"excerpt":"Translated by ChatGPT.","content":"\nLet's explain this using the directory structure provided on the official website:\n\n```\nsound/ Top-level package \n    init.py Initialize the sound package \n    formats/ Subpackage for file format conversions \n        init.py \n        wavread.py \n        wavwrite.py \n        aiffread.py \n        aiffwrite.py \n        auread.py \n        auwrite.py \n        ... \n    effects/ Subpackage for sound effects \n        init.py \n        echo.py \n        surround.py \n        reverse.py \n        ... \n    filters/ Subpackage for filters \n        init.py \n        equalizer.py \n        vocoder.py \n        karaoke.py \n        ...\n```\n\n## Using Existing Python Code\n\nMost programming tutorials start teaching syntax after setting up the environment. But let’s first discuss using code that’s already been written. The simplest way to do this is by installing packages through a package manager:\n\n```bash\npip install sound\n```\n\nThen, if you want to use a function in a specific file (let's say `wavwrite.py` has a function called `write()`), there are multiple ways to import it. Note the different ways to import and call the function:\n\n```python\nimport sound\nsound.formats.wavwrite.write()\n\nfrom sound import formats\nformats.wavwrite.write() \n\nfrom sound.formats import wavwrite\nwavwrite.write()\n\nfrom sound.formats.wavwrite import write\nwrite()\n```\n\nHowever, not all Python code can be installed directly. For example, if a research paper is published along with data processing code, often the code is simply shared in a folder. If you download that folder and try `import sound`, an error will occur, stating that Python can't find a library named sound.\n\n## How Python Locates Code Files\n\nIf you think about it, it’s normal that an error occurs. The previous simple command `import sound` solved the problem magically—finding the library without needing a file path—even though libraries can be located in different parts of the file system.\n\nThis is because Python doesn’t search the entire hard drive. Instead, it uses a variable, typically named `PYTHONPATH`, which holds a list of directories where Python library folders are located. When you enter a command in the terminal, Python will:\n\n- Search the current folder, meaning the folder where you launched the Python command.\n- Look through each folder in the `PYTHONPATH` variable.\n- Finally, check the default location for packages installed by the package manager, typically `\u003cpath to python\u003e/site-packages`.\n\nIf the library is found, Python will import it; otherwise, an error occurs.\n\nIn the error mentioned above, if we happen to be in the `sound` folder and run Python, the first condition applies, and `import sound` won’t produce an error. But this won’t work in other locations.\n\n## Terminology: Interactive, Script, Module, Package\n\nExecutable Python commands can be found in the following contexts, with the first being interactive and the last three being file-based:\n\n1. interactive: The Python **interactive interface**, also known as calculator mode, which appears after entering `python` in the command line. Commands entered here display results immediately. Once Python exits, these commands are lost.\n2. script: A Python **script** file, like `somefile.py` when running `python somefile.py` from the command line.\n    1. Python is lightweight, and some commands are worth saving for later reuse, much like Linux commands and bash scripts.\n    2. However, as a fully-featured language, Python can also support complex object-oriented programming. In this sense, the main script file can be considered the main module, serving as the main file and program entry point.\n3. module: A Python **module** file, which you can call with `python -m another` (without a `.py` extension). Technically, any `.py` file is a module, but it’s generally used for files whose variables and functions are meant for other Python files to import.\n4. package: A Python **package**, which is a higher-level structure of related modules (typically a folder with `__init__.py` inside). But packages and modules aren’t defined solely by folder hierarchy—more on that in the next section.\n\n## Script vs. Module\n\nPython offers both scripting flexibility and the robustness of more complex languages. Unlike some languages, you don’t need a main file called `main.py` with a `Main` class and a `main()` function. But Python still needs a starting point in a complex program.\n\nThis starting point is the `.py` file after `python` (without the `-m` parameter). Python sets the `__name__` attribute of this main file to `\"__main__\"`. This allows Python to treat it as the main script file, even if it's part of a larger library, with no knowledge of sibling or other related modules.\n\nOther modules are used by importing them in the main module (or script). The import command tells Python where to find the module’s code and the relationships between modules in the package hierarchy. Alternatively, the `-m` option provides this structure, e.g., `python -m sound.formats.wavwrite`, which allows `wavwrite.py` to execute and know its relative position in the hierarchy.\n\n## Absolute Import vs. Relative Import\n\nThe examples at the beginning use absolute imports (no leading `.`), whereas relative imports (`.` for current directory, `..` for parent directory) are for internal code that's almost never run as the main module.\n\nReturning to the file structure above, if `sound/effects/surround.py` needs functions from `sound/formats/wavwrite.py` and `sound/effects/echo.py`, you could write:\n\n```python\n# in sound/effects/surround.py\nfrom ..formats import wavwrite\nfrom . import echo\n```\n\n## Organizing Code for Reuse\n\nFinally, after years of coding, I’m drafting a paper, and now I need to organize previous analysis code into a project. Each time I tried to unify everything, import errors between modules were an issue. This post resulted from that experience.\n\nHere's a recommended project structure from [this article](https://gist.github.com/ericmjl/27e50331f24db3e8f957d1fe7bbbe510):\n\n```\n|- notebooks/\n   |- 01-first-logical-notebook.ipynb\n   |- 02-second-logical-notebook.ipynb\n   |- prototype-notebook.ipynb\n   |- archive/\n      |- no-longer-useful.ipynb\n|- projectname/\n   |- projectname/\n      |- __init__.py\n      |- config.py\n      |- data.py\n      |- utils.py\n   |- setup.py\n|- README.md\n|- data/\n   |- raw/\n   |- processed/\n   |- cleaned/\n|- scripts/\n   |- script1.py\n   |- script2.py\n   |- archive/\n      |- no-longer-useful.py\n|- environment.yml\n```\n\nThis structure allows you to convert tools into an installable library. Now, you can use `projectname` in notebooks and scripts just like `numpy` or `pandas`.\n\nTo use the code, install it with:\n\n```bash\ncd projectname\npip install -e .\n```\n\n## References\n\n- [What’s the difference between a Python module and a Python package?](https://stackoverflow.com/questions/7948494/whats-the-difference-between-a-python-module-and-a-python-package) All Python files are modules, while a package is a specific kind of module.\n- Official explanation of Python modules: https://docs.python.org/3/tutorial/modules.html\n- [StackOverflow: \"run as module\" is different from \"run as script\".](https://stackoverflow.com/questions/14132789/relative-imports-for-the-billionth-time) Running as a module sets the `__name__` attribute to the module’s name, while running as a script sets it to `__main__`.\n- A project organization tutorial: https://realpython.com/python-application-layouts/\n- Official documentation on packaging: https://packaging.python.org/en/latest/tutorials/packaging-projects/\n"},{"slug":"install-pytorch-cpu-on-fedora","filename":"2021-10-13-install-pytorch-cpu-on-fedora.md","date":"2021-10-13","title":".py | Installing the CPU Version of PyTorch on Fedora","layout":"post","keywords":["md","py","ai"],"excerpt":"Translated by ChatGPT.","content":"\nI'm about to attend a summer school on deep learning applications in microscopic image processing.\n\nDeep learning is a branch of machine learning, where most data can be represented as vectors (first-order tensors) and most algorithms can be broken down into operations or transformations on vectors, represented as matrices (second-order tensors). This need for efficient tensor computation led to libraries like PyTorch, TensorFlow, and others such as Keras, Caffe, and more. In earlier versions, PyTorch and TensorFlow differed significantly, but over time, they've merged in features and have edged out other competitors; they're starting to look increasingly similar. \"They've become what they once hated.\" lol\n\nI'm not sure which machine learning framework will be used in the course, so I've decided to install both PyTorch and TensorFlow (shrug). This is also a continuation of my previous [Python tutorial](python-interpreter-editor-virtualenv).\n\nHere’s my software and hardware setup: an Intel CPU with integrated graphics (can't run *Civilization VI*). Even though my system isn't officially supported on Linux, I’ve managed to get Fedora running smoothly after a few kernel updates. I'm using Python version 3.9.6, pip as my package manager, and VS Code as my editor.\n\n## Creating a Virtual Environment\n\nI explained why to create a virtual environment in the previous post, so let's jump right in. I created two virtual environments, naming them `torch` and `tf`. In the command line snippets, I've added a prompt `(env)[me@mycomputer]$` to show which environment I'm in—remove this prompt when copying the commands.\n\n```bash\n[me@mycomputer]$ mkvirtualenv torch\n(torch)[me@mycomputer]$\n```\n\n## Installation\n\nOn the PyTorch website, find the installation command for your setup: https://pytorch.org/get-started/locally/. In my case, I chose `Stable` \u003e `Linux` \u003e `Pip` \u003e `Python` \u003e `CPU`. Copy the generated command and paste it into the terminal:\n\n```bash\n(torch)[me@mycomputer]$ pip3 install torch==1.9.1+cpu torchvision==0.10.1+cpu torchaudio==0.9.1 -f https://download.pytorch.org/whl/torch_stable.html\n```\n\nWait for the installation messages to confirm success.\n\n## Verification and Deactivation\n\nVerify the installation using the [official guide](https://pytorch.org/get-started/locally/#linux-verification):\n\n```python\nimport torch\nx = torch.rand(5, 3)\nprint(x)\n\n# tensor([[0.3799, 0.4494, 0.4296],\n#       [0.5800, 0.0180, 0.3110],\n#       [0.9847, 0.0125, 0.2648],\n#       [0.0296, 0.3142, 0.9266],\n#       [0.3192, 0.9645, 0.5545]])\n```\n\nNext, deactivate the environment to prepare for installing TensorFlow:\n\n```bash\n(torch)[me@mycomputer]$ deactivate\n[me@mycomputer]$\n```\n\n## What About TensorFlow?\n\nInitially, I planned to cover both PyTorch and TensorFlow in this post, but TensorFlow proved a bit more challenging.\n\n### Direct Installation\n\nGo to the [TensorFlow website](https://www.tensorflow.org/install) and click the \"Install\" button in the top navigation. Then, on the left sidebar, find the package/pip section; the [installation command](https://www.tensorflow.org/install/pip#3.-install-the-tensorflow-pip-package) is just one line:\n\n```bash\npip install --upgrade tensorflow\n```\n\nHowever, this command didn’t work as expected. The installation completed without errors, but running a test produced numerous error messages.\n\nThe error log included `Could not load dynamic library 'libcudart.so.11.0'`, suggesting that this command installed the GPU version by mistake.\n\n### Installing the CPU Version\n\nI found the CPU version download link in the \"Package Location\" section of the page: `https://storage.googleapis.com/tensorflow/linux/cpu/tensorflow_cpu-2.6.0-cp39-cp39-manylinux2010_x86_64.whl`. I deleted the virtual environment, created it again, and reinstalled TensorFlow:\n\n```bash\n(tf)[me@mycomputer]$ deactivate\n[me@mycomputer]$ rmvirtualenv tf\n[me@mycomputer]$ mkvirtualenv tf\n(tf)[me@mycomputer]$ pip install --upgrade pip\n(tf)[me@mycomputer]$ pip install https://storage.googleapis.com/tensorflow/linux/cpu/tensorflow_cpu-2.6.0-cp39-cp39-manylinux2010_x86_64.whl\n```\n\nAfter testing with the official command, I still got warnings:\n\n```bash\n(tf) [shixing@yoga-laptop ~]$ python -c \"import tensorflow as tf;print(tf.reduce_sum(tf.random.normal([1000, 1000])))\"\n20XX-XX-XX XX:XX:XX.XXXXXX: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA\nTo enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\ntf.Tensor(-1338.4773, shape=(), dtype=float32)\n```\n\nThis Stack Overflow answer mentions that to avoid this warning, TensorFlow needs to be compiled from source. The official guide to building from source provides more details, but this is quite a hassle, so I’ll cover it in another post (to be continued).\n"},{"slug":"python-interpreter-editor-virtualenv","filename":"2021-06-29-python-interpreter-editor-virtualenv.md","date":"2021-06-29","title":".py | Python's Interpreter, Editor, and Virtual Environment","layout":"post","keywords":["md","py"],"excerpt":"Really, what are Python's interpreter, editor, and virtual environment","content":"\n\nIn the previous post \"[Configuring Python Development Environment on Windows 10](python-installation-and-configuration)\", I did roughly three things:\n- First, I installed the Python interpreter.\n- Second, I installed VSCode on the computer.\n- Third, as we know, is the virtual environment, virtualenv.\n\nIf there's something else worth mentioning, it's to add a bit of \"what\" alongside the \"how\". To make the article flow better, I plan to rearrange the order: first, talk about the editor, then the interpreter, and finally, the virtual environment. Although the article is based on Python, these concepts apply to almost all general-purpose programming languages. Because I'm not from a computer science background, this article is just my study notes, so if there are any inaccuracies, feel free to point them out.\n\n## Editor\n\nFirst, let's do an experiment. Rename the Python file `hello.py` created in the previous tutorial, change the extension `.py` to `.txt`, then double-click to open the file. What happens? In the Windows system, the most ordinary Notepad window pops up, displaying Python statements in black text on a white background, unlike in VSCode where different statements have different colors, but the content is exactly the same.\n\nFurthermore, you can directly execute the content of a `.txt` file in the command line with the Python interpreter (`py -m hello.txt`), with the same effect as `.py`.\n\nThe `.py` extension doesn't do anything special; like a `.txt` file, its content is simply a string of characters readable by humans. Such files are called [text files](https://en.wikipedia.org/wiki/Text_file). An editor is a program that handles text files. Notepad is one type of editor, as is VSCode.\n\nSince Notepad can already write Python code, why bother installing VSCode? Indeed, some people really only use Notepad or other built-in editors of the operating system to write Python code (there's even a video on YouTube of someone writing code in Microsoft Word); then debug the program using the command line. However, VSCode is specifically developed for programmers, providing many features that default editors lack, such as syntax highlighting, which highlights functions, variables, and keywords in different colors.\n\n## Compiler / Interpreter\n\nIn addition to text files, the second type of file mentioned in this article is a [binary file](https://en.wikipedia.org/wiki/Binary_file).\n\nComputers cannot directly understand characters recognized by humans. At the lowest level, classic computers understand only 0s and 1s represented in different ways. Although all files are ultimately binary at the lowest level, the term \"binary file\" is generally used specifically to refer to files other than text files. Since image files, video files, etc., have their own names, the files referred to by this term are basically related to software programs, files that can be executed by computers.\n\nHumans recognize characters, computers recognize only 0s and 1s, so the most intuitive approach is to translate text files into binary files. This translation process is called [compilation](https://en.wikipedia.org/wiki/Compiler), and the software that can accomplish this is called a compiler. Once compilation is complete, the compiler exits, and to execute the program, the computer simply executes the compiled executable file.\n\nHowever, compilation has a problem: the entire software needs to be compiled into a complete executable with all source code text files written. As the scale of the software grows, the resources and time required for compilation increase; once a modification is made, the entire project needs to be recompiled. Many times, we only want to quickly know the effect of a command in a large project; compilation is not suitable for such scenarios.\n\nThus, the [interpreter](https://en.wikipedia.org/wiki/Interpreter_(computing)) emerged. This type of program is much more complex than a compiler. It runs whenever we execute a command in a programming language, allowing us to enter commands one by one, remembering the context of the code, and even remembering the results of our previous commands. The cost is that even for large projects, it needs to analyze and interpret line by line, and its computational resource overhead and speed are not as efficient as compilation.\n\nPython is a (official implementation) programming language that uses an interpreter. The `python-***.exe` file we download from the official website is the Python interpreter. This also means that Python programs tend to perform less well than programs written by C/C++ programmers of the same level. However, due to its suitability for trial and error with single-line execution, Python is quite popular in research fields.\n\nOf course, Python's characteristics are far more than just being an interpreted language. It is also:\n- [Scripting language](https://program-think.blogspot.com/2009/08/why-choose-python-1-script.html)\n- [Dynamic language](https://program-think.blogspot.com/2009/08/why-choose-python-2-dynamic.html)\n- [Object-oriented language](https://program-think.blogspot.com/2010/08/why-choose-python-3-oop.html)\n- [Functional programming language](https://program-think.blogspot.com/2010/08/why-choose-python-3-oop.html)\n- [...](https://program-think.blogspot.com/2010/08/why-choose-python-5-tools.html)\n\n## Virtual Environment\n\nOnce I asked my girlfriend what IDE she uses to write Python, and she proudly replied that her MacBook comes with Python, so she can run it directly from the command line... The problem wasn't that she didn't answer my question directly (actually, thinking about it, she did answer part of the question), but rather that running the system's built-in Python interpreter or other programming language interpreters or compilers directly from the command line is a dangerous practice common among beginner programmers.\n\nFortunately, Windows is fine; after all, it's an operating system aimed at the general consumer, so it doesn't come with Python pre-installed. However, for more geek-oriented operating systems like Linux and the BSD family, things are different. These operating systems (or their distributions) often come pre-installed with Python. This Python isn't intended for users to develop their own programs but rather for the many system tools written in Python to call upon. Since the version of Python is usually controlled by the package manager of the distribution, it often lags behind the latest Python version for some time to avoid bugs in the new version and to give developers of system tools time to update their code. So, if you use this version of Python for development and accidentally upgrade Python, it's likely to cause some system functionality to malfunction.\n\nA virtual environment is Python's solution to this problem. We can install a version of Python different from the system's default version but not add this interpreter to the system path, so the operating system is unaware of the existence of this version of Python. When creating a virtual environment, we specify the use of this specific version of Python. So, after activating the virtual environment, it becomes the Python we need for development, and exiting the virtual environment returns to the Python used by system tools.\n\nMoreover, even if two projects use the same version of Python and both use the system's default version, virtual environments still have a place. Most programs require dependency onlibrary code written by others, called libraries. Different projects may depend on different library code, or different versions of the same library. In this case, you can create separate virtual environments and install the libraries corresponding to each project. Projects can be isolated from each other.\n\n## Editor + Interpreter / Compiler + Virtual Environment Management + ... = IDE\n\nIt's too cumbersome to install so many different programs for Python development. Can't it be done with just one click? Of course it can. A program that integrates various tools used in the development process is called an [integrated development environment (IDE)](https://en.wikipedia.org/wiki/Integrated_development_environment). For Python, the most famous IDE is undoubtedly [Anaconda](https://www.anaconda.com/).\n\nThen why don't I use it? Of course, I've used it before, but after hearing about the fame of VSCode and trying it out, I don't want to go back. Writing Python requires editing `.py` files, writing blog posts requires editing `.md` files, and some blog features require JavaScript. Essentially, all these tasks involve editing text files. In an editor like VSCode, everything is done together. So, programs like Spyder and Jupyter Lab in Anaconda seem redundant.\n\n\"Doing one thing well\" is also part of the [Unix philosophy](https://en.wikipedia.org/wiki/Unix_philosophy). But the problem is that different people have different definitions of \"one thing\". Some people think making breakfast is one thing, while others think it involves several things like heating milk, frying eggs, and toasting bread. Who is right?\n\nPerhaps both are right, but Editor + Interpreter + ... is upstream on the chain of contempt compared to IDE. You may disagree with this, but you should know it; otherwise, when others show off, you might feel embarrassed for not being prepared.\n"},{"slug":"mpipks-non-equilibrium-physics-note-10-dimension-reduction","filename":"2021-05-15-mpipks-non-equilibrium-physics-note-10-dimension-reduction.md","date":"2021-05-15","title":".mpipks-note | 10. Dimension Reduction","layout":"post","keywords":["mpipks-note"],"excerpt":"As the title suggests","hasMath":true,"content":"\n\u003e 第十讲是由校外邀请来的嘉宾来讲数据降维，所以课程网站上没有课件。时间轴和语音转录相对来说不那么有意义了。所以在听课的过程中做了如下笔记。\n\n## Contents\n\n- High dimensinoal data\n- Dimensionality Reduction Methods\n- Clustering (a discrete form of reduction)\n\n## High Dimensional Data\n\ndata matrix X:n*m. \n\nn samples as rows, m columns as observables\n\nHow to visualize?\n- m=1: list\n- m=2: x-y plot\n\n__Aim of dimension reduction: reduce the number of observables, while keeping relevant information.__\n\n### Examples\n\nFerromagnetic Ising model of a 28*28 lattice, 50,000 samples. The 1st principal component scales with the magnetization. \n\n500,000 DNA sites of 3,000 humans. Projections onto the first 2 principal components givees a rough map of Europe\n\nFashion MNIST: 60,000 pictures of 784 (28*28) pixels. UMAP projections onto 2 dimensions \n\n## Linear dimension reduction\n\n- Principal component analysis\n- Non-negative matrix factorization\n- Independent component analysis\n- Factor analysis\n\n### Principal component analysis\n\nchange of basis\n\n__Aim:__ to find a transformation such that\n- the components are uncorrelated\n- ordered by the explained variability\n\n__Math__:\n\n$$ X^T X $$ is a $$m \\times m$$ matrix. Eigen-decomposition gives $$(\\lambda,W) $$ where $$W$$'s columns are eigen vectors sorted by $$\\lambda$$\n\n$$ T = X \\cdot W $$ where  $$T$$ is the score onto principal components, and $$T^TT$$ is diagonal\n\n$$ T_r = X \\cdot W_r $$ where $$W_r$$ is a n*r matrix of the first r components\n\n__Practice In `R`__: \n\n- Iris Dataset\n- Pitures of Objects at Different Angles: \n- Swiss Roll: the transformed data almost have the same structure as the original data. This shows the limit of PCA due to linearity\n\n[wikipedia - Nonlinear_dimensionality_reduction]()\n\n### UMAP: uniform manifold approximation\n\n__Assumptions__:\n- uniformed distributeed on Riemann manifold\n- Riemann metric is locally constant\n- The manifold is locally conncected\n\n__Results__:\n- models the manifold with a fuzzy topological structure\n- finds low dimensional projection of the data that has the cloest possible equivalent fuzzy topological structure \n\n__Simplex__\n0-simplex: dot\n1-simplex: line\n2-simplex: triangle\n3-simplex: 正四面体\n\n__Problem__:\n\nWhen the uniform distribution is not met, the result may not be what we like. To circumvent this, we need to define the local metric with the nearest neighbours.\n\n__Finding a low dimensional representation__\n- derive similar fuzzy topological representation of a low dimensional representation\n- interpret edge weight s as proba\n- minimize cross entropy\n- [pair-code.github.io/understand_umap](pair-code.github.com/understand_umap)\n\n__Practice in `R`__\n- Swiss Roll becomes a 2-d ring\n- Pictures at different angles are well seperated\n\n## Clustering\n\n### hierarchiral clustering\n- metric\n- linkage criterion / distance between clusters\n\n### centroid base clustering (k-means)\n- pre-specify number of clusters, k\n- partition the data such that the squared distance to cluster mean position is minimal\n- Lloyd algorithm\n    - start with random centers\n    - assign data to neatest center\n    - compute new centers\n\n### graph based-clustering\n- graph representation of data\n- partition the data such that modularity Q is optimized\n- $$ Q = \\sum_{i=1}^c e_{ii}-\\gamma {a_i}^2, a_i = \\sum_{j-1}^c e_{ij} $$.\n    - $$ e_{ij} $$: fraction of links between cluster i and j\n    - $$ \\gamma $$: resulution\n\n## Summary\n\nQuestion: How to identify order in high-dimensinoal data?\n- dimensionality reductuon to visualize high-dimensional data and to identify macroscopic observables\n- clustering to identify discrete order in high dimensional data\n- methods to se depending on the problem at hand\n"},{"slug":"python-installation-and-configuration","filename":"2021-03-07-python-installation-and-configuration.md","date":"2021-03-07","title":".py | Setup a Python Dev Environment","layout":"post","keywords":["md"],"excerpt":"Setup a Python dev environment on a Windows machine, using barebone Python, virtualenvwrapper, vscode, and without Conda.","content":"\nMy Python version has been stuck at 3.7 for a while, but now the latest version is already at 3.9. Recently, I needed to study a piece of Python code written by someone else, who specifically requested Python version 3.6. So, I uninstalled the 3.7 version and reinstalled both Python versions 3.9 and 3.6.\n\nI've been trying to encourage my parents to learn programming for a long time, and I owe them a tutorial. Whether it's for those who are committed to continuous improvement in their first programming language or for those who intend to dabble lightly in programming, Python is an excellent choice. So, even though my parents have no interest in programming whatsoever, I still completed this tutorial, hoping it could help others. Since it's aimed at complete beginners, it's based on the Windows 10 operating system. And because I also use this setup myself, based on personal preference, I didn't directly use `conda`.\n\n## Table of Content\n- python\n    - Download `python` interpreter from the official website\n    - Install `python` interpreter\n- Virtual Environment\n    - Set environment variable\n    - Install `virtualenv` with `pip`\n    - Install `virtualenvwrapper` with `pip`\n    - Create a virtual env, and install `jupyter`\n- Editor/IDE\n    - Download `vscode` from official website\n    - Install `vscode` and related plugins\n    - Verify it works\n\n## Let's go\n\n### Download `python` interpreter from the official website\n\nDownload from the official website of Python ([https://www.python.org/downloads/](https://www.python.org/downloads/)).\n\n![](/blog/photos/2021-03-07-python-official-website.png)\n\n### Install `python` interpreter\n\nAfter executing the downloaded file, you'll see the interface as shown in the screenshot below. Click on the option highlighted in red box because I want to change the installation path. It seems that you can also check the \"Add Python 3.9 to PATH\" at the bottom. I didn't select it during installation, but later I found out that this step needs to be done manually.\n\n![](/blog/photos/2021-03-07-python-install.png)\n\nFollowing that, check each option in the next interface, then click \"Next\".\n\n![](/blog/photos/2021-03-07-python-next.png)\n\nIn the upcoming interface, select \"Install for all users\" and then choose an installation path you prefer. The reason I choose C:\\Python39 for this location is because:\n\n1. The C drive is the system drive and it's a solid-state drive on my machine, which is faster.\n2. The default path is in the personal folder, which is quite hard to find.\n3. `C:\\Program Files` requires administrator permissions, which is inconvenient for using pip for package management.\n\n![](/blog/photos/2021-03-07-python-install-path.png)\n\nAfter the installation is complete, open the command prompt and type `py`. You should see a string of Python version information, and then the command prompt will change to `\u003e\u003e\u003e`, indicating that you have entered the Python environment, where you can execute Python commands line by line. Type `exit()` to exit Python.\n\n### Set environment variable\n\nIn the Windows search bar, type \"environment variables,\" and before you finish typing, you should see the suggested result as shown in the image below:\n\n![](/blog/photos/2021-03-07-environment-variable-search.png)\n\nThen, follow the steps in the image below to complete two tasks:\n1. Add two new values, `C:\\Python39` and `C:\\Python39\\Scripts`, to the `Path` variable.\n2. Create a new folder for storing Python virtual environments (such as `C:\\PythonEnvs`). Create a new variable named `WORKON_HOME` and set its value to `C:\\PythonEnvs`.\n\n![](/blog/photos/2021-03-07-environment-variable-box.png)\n\n### Install `virtualenv` and `virtualenvwrapper` with `pip`\n\nFollowing the steps just now, `pip` should now be installed. You should see similar feedback in the command line after entering the commands along the way:\n\n![](/blog/photos/2021-03-07-pip-check.png)\n\nIf the command line indicates that `pip` is not found, you may need to reinstall it using the method outlined in [this link](https://pip.pypa.io/en/stable/installing/#installing-with-get-pip-py).\n\nType `py -m pip install virtualenv` in the command line to install the virtual environment manager:\n\n![](/blog/photos/2021-03-07-virtualenv-install.png)\n\nAfter a successful installation message appears, type `py -m pip install virtualenvwrapper-win` in the command line. The result should be similar to the image above; I forgot to take a screenshot.\n\n### Create a virtual env, and install `jupyter`\n\nAfter installing `virtualenv` and `virtualenvwrapper`, enter `mkvirtualenv base` in the command line to create a new virtual environment named base.\n\nThen enter `workon base`. At this point, you should see `(base)` at the beginning of the command prompt, indicating that we are now working within the base virtual environment. Now, when you check `pip --version`, you'll notice that the Python location has changed from before.\n\n![Example Image](/blog/photos/2021-03-07-virtualenvwrapper-check.png)\n\nWithin this environment, install Jupyter by typing `pip install jupyter`. While this step is not mandatory, it allows us to have a similar experience to Spyder, which comes with Conda, once we install VSCode. This makes it very convenient for beginners to debug line by line.\n\n### Download and install `vscode`\n\nDownload the installation package from the official website of VS Code ([link here](https://code.visualstudio.com/)), where you'll find a prominent download button:\n\n![](/blog/photos/2021-03-07-vscode-download.png)\n\nAfter downloading, the installation process is similar to that of any other program (it's been a while, so no screenshots).\n\nOnce installed, open the program and click on the icon outlined in red on the left toolbar, then search for Python-related VS Code plugins. For Python development, you generally need `python`, `pylance`, and `jupyter`:\n\n![](/blog/photos/2021-03-07-vscode-plugin.png)\n\nIn the future, when you want to run VS Code, right-click in the folder where you want to start coding and select \"Open with Code\":\n\n![](/blog/photos/2021-03-07-vscode-folder-open.png)\n\nNext, in the VS Code interface, click on the button outlined in red in the file tree to create a new file, name it `*.py`, and here I just wrote something like \"hello.\"\n\n![](/blog/photos/2021-03-07-vscode-new-file.png)\n\nIn the subsequently opened document editing area, type `print(\"Hello World\")`, place the cursor on this line, and press `Shift`+`Enter` to execute this line:\n\n![](/blog/photos/2021-03-07-vscode-run.png)\n\nAs it's the first time executing and since `jupyter` is installed, you should see a prompt in the bottom right corner of VS Code asking if you want to execute the code using the Jupyter interactive window. Then close the terminal below (the cross in the top right corner of the bottom half). Press `Shift`+`Enter` again to execute. At this point, the result should look like the image below. Click on the variable explorer in the red circle, and it's almost the same experience as Spyder.\n\n![](/blog/photos/2021-03-07-vscode-interactive.png)\n\nIn the future, the Jupyter Interactive Window could be called up by typing `Ctrl` + `Shift` + `P` and searching \"interactive\" in the command line that appears on top of the screen.\n"}]},"__N_SSG":true},"page":"/en/history","query":{},"buildId":"buildID","isFallback":false,"gsp":true,"scriptLoader":[]}</script></body></html>